{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[Biopython](www.biopython.org) is a well-maintained collection of modules, tools, and scripts specifically designed to help with common bioinformatics tasks.  It is free and open source, with a large number of contributors maintaining the package. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequences and Sequence objects\n",
    "Earlier, we created and manipulated DNA sequences as strings. A core component of biopython is the creation of specific 'classes' to store and manipulate certain types of items. One of these is the `Seq` class which is designed to hold all of the relevant information needed for biological sequences (DNA, RNA, protein, etc). There are two primary components for a biological sequences.  The sequence itself and the 'Alphabet' used to define what the characters in the sequence represent (i.e. [A,G,C,T] for DNA or [A,G,C,U] for RNA).\n",
    "\n",
    "Lets start by creating a 'generic' sequence (`Seq` object) in biopython."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTGAGCAGTTGAGCGATTATGTCGATTAGCTGAGGCATTCGAGCTATG\n",
      "<class 'Bio.Seq.Seq'>\n",
      "Alphabet()\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq # first we import a function from a sub module of the Bio (biopython) module directly.\n",
    "\n",
    "mySeq = Seq(\"TTGAGCAGTTGAGCGATTATGTCGATTAGCTGAGGCATTCGAGCTATG\")\n",
    "\n",
    "print(mySeq)\n",
    "print(type(mySeq))\n",
    "\n",
    "#str(mySeq)\n",
    "\n",
    "print(mySeq.alphabet)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So why go to all of this trouble to create a `Seq` object instead of just storing the DNA seqence as a string?  Well, the `Seq` object defines a number of methods (functions) that code for useful _biological_ operations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTGAGCAGTTGAGCGATTATGTCGATTAGCTGAGGCATTCGAGCTATG\n",
      "AACTCGTCAACTCGCTAATACAGCTAATCGACTCCGTAAGCTCGATAC\n",
      "CATAGCTCGAATGCCTCAGCTAATCGACATAATCGCTCAACTGCTCAA\n",
      "UUGAGCAGUUGAGCGAUUAUGUCGAUUAGCUGAGGCAUUCGAGCUAUG\n",
      "LSS*AIMSIS*GIRAM\n",
      "5\n"
     ]
    }
   ],
   "source": [
    "print(mySeq)\n",
    "print(mySeq.complement())\n",
    "print(mySeq.reverse_complement())\n",
    "print(mySeq.transcribe())\n",
    "print(mySeq.transcribe().translate())\n",
    "print(mySeq.count(\"TT\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Seq` objects store sequence information `SeqRecord` objects store metadata information associated with an individual `Seq` object including an identifier (id) a name, and a description.  We'll see more of these as we progress."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsers\n",
    "An important task in bioinformatics is 'parsing' files of different format. There are many types of file formats for biological data, some of which are human readable, and some are not readily so. And some of them can contain thousands or millions of data points that you may need to iterate over in order to analyze. Thankfully biopython maintains several functions in the `Bio.SeqIO` module to parse common biological file formats.\n",
    "\n",
    "You've already been introduced to the FASTA format earlier today. FASTA files can contain many unique sequences within a single file. Using biopython, we can parse a FASTA file into individual `SeqRecord` objects and then iteratively do something with them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGGCGTTGTCTTTAATTCGATTAGCATTGCATATCGATTATCTAGCGATATGCTATGCTTAGC\n",
      "AGCCATATGTATCGAGCGATATGCGATACGAGTATCGAGTATGCAGTATGCATTGCAGTATGC\n",
      "TGGGTATGCGCGACGAGCATTATACGATGCATTATTACGGATCTACGGCGATATTACGTACGA\n",
      "GGACGTATCGAGTCTAGCGAGCGACTTCGAGCGATATCGGACTCTCGTCCTCTTCAGTCAGCC\n",
      "TTTACGGACTTACGGACTAGCTGAGCTAGCTACGATCGATCGATCGTAGCTACGATCGTAGCT\n",
      "CATCGATCGTAGC\n",
      "ACACGGACTAGCGGATCTATCTGTACTGAGCGTATCTGACGGTAGCTATCGGACGTATCGGACGGACACAGCGTATGCGAC\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO # notice we are importing a different sub module called SeqIO\n",
    "\n",
    "fasta_file = SeqIO.parse(\"test.fa\",'fasta')\n",
    "\n",
    "# Lets access the `Seq` object contained within each returned `SeqRecord` and print that out.\n",
    "for record in fasta_file:\n",
    "    print(record.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When we use this and other parsers, they will return an object of a relevant class.  In the case of `SeqRecord` objects, there are additional properties parsed and stored"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: dna_sequence_1\n",
      "Name: dna_sequence_1\n",
      "Description: dna_sequence_1\n",
      "Number of features: 0\n",
      "Seq('TGGCGTTGTCTTTAATTCGATTAGCATTGCATATCGATTATCTAGCGATATGCT...AGC', DNAAlphabet())\n",
      "dna_sequence_1\n",
      "63\n",
      "dna_sequence_1\n",
      "Help on SeqRecord in module Bio.SeqRecord object:\n",
      "\n",
      "class SeqRecord(builtins.object)\n",
      " |  A SeqRecord object holds a sequence and information about it.\n",
      " |  \n",
      " |  Main attributes:\n",
      " |   - id          - Identifier such as a locus tag (string)\n",
      " |   - seq         - The sequence itself (Seq object or similar)\n",
      " |  \n",
      " |  Additional attributes:\n",
      " |   - name        - Sequence name, e.g. gene name (string)\n",
      " |   - description - Additional text (string)\n",
      " |   - dbxrefs     - List of database cross references (list of strings)\n",
      " |   - features    - Any (sub)features defined (list of SeqFeature objects)\n",
      " |   - annotations - Further information about the whole sequence (dictionary).\n",
      " |     Most entries are strings, or lists of strings.\n",
      " |   - letter_annotations - Per letter/symbol annotation (restricted\n",
      " |     dictionary). This holds Python sequences (lists, strings\n",
      " |     or tuples) whose length matches that of the sequence.\n",
      " |     A typical use would be to hold a list of integers\n",
      " |     representing sequencing quality scores, or a string\n",
      " |     representing the secondary structure.\n",
      " |  \n",
      " |  You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |  SeqRecord objects.  However, you may want to create your own SeqRecord\n",
      " |  objects directly (see the __init__ method for further details):\n",
      " |  \n",
      " |  >>> from Bio.Seq import Seq\n",
      " |  >>> from Bio.SeqRecord import SeqRecord\n",
      " |  >>> from Bio.Alphabet import IUPAC\n",
      " |  >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\",\n",
      " |  ...                         IUPAC.protein),\n",
      " |  ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |  ...                    description=\"toxic membrane protein\")\n",
      " |  >>> print(record)\n",
      " |  ID: YP_025292.1\n",
      " |  Name: HokC\n",
      " |  Description: toxic membrane protein\n",
      " |  Number of features: 0\n",
      " |  Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF', IUPACProtein())\n",
      " |  \n",
      " |  If you want to save SeqRecord objects to a sequence file, use Bio.SeqIO\n",
      " |  for this.  For the special case where you want the SeqRecord turned into\n",
      " |  a string in a particular file format there is a format method which uses\n",
      " |  Bio.SeqIO internally:\n",
      " |  \n",
      " |  >>> print(record.format(\"fasta\"))\n",
      " |  >YP_025292.1 toxic membrane protein\n",
      " |  MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  <BLANKLINE>\n",
      " |  \n",
      " |  You can also do things like slicing a SeqRecord, checking its length, etc\n",
      " |  \n",
      " |  >>> len(record)\n",
      " |  44\n",
      " |  >>> edited = record[:10] + record[11:]\n",
      " |  >>> print(edited.seq)\n",
      " |  MKQHKAMIVAIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  >>> print(record.seq)\n",
      " |  MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Add another sequence or string to this sequence.\n",
      " |      \n",
      " |      The other sequence can be a SeqRecord object, a Seq object (or\n",
      " |      similar, e.g. a MutableSeq) or a plain Python string. If you add\n",
      " |      a plain string or a Seq (like) object, the new SeqRecord will simply\n",
      " |      have this appended to the existing data. However, any per letter\n",
      " |      annotation will be lost:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      \n",
      " |      >>> new = record + \"ACT\"\n",
      " |      >>> print(\"%s %s\" % (new.id, new.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNNACT\n",
      " |      >>> print(list(new.letter_annotations))\n",
      " |      []\n",
      " |      \n",
      " |      The new record will attempt to combine the annotation, but for any\n",
      " |      ambiguities (e.g. different names) it defaults to omitting that\n",
      " |      annotation.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> with open(\"GenBank/pBAD30.gb\") as handle:\n",
      " |      ...     plasmid = SeqIO.read(handle, \"gb\")\n",
      " |      >>> print(\"%s %i\" % (plasmid.id, len(plasmid)))\n",
      " |      pBAD30 4923\n",
      " |      \n",
      " |      Now let's cut the plasmid into two pieces, and join them back up the\n",
      " |      other way round (i.e. shift the starting point on this plasmid, have\n",
      " |      a look at the annotated features in the original file to see why this\n",
      " |      particular split point might make sense):\n",
      " |      \n",
      " |      >>> left = plasmid[:3765]\n",
      " |      >>> right = plasmid[3765:]\n",
      " |      >>> new = right + left\n",
      " |      >>> print(\"%s %i\" % (new.id, len(new)))\n",
      " |      pBAD30 4923\n",
      " |      >>> str(new.seq) == str(right.seq + left.seq)\n",
      " |      True\n",
      " |      >>> len(new.features) == len(left.features) + len(right.features)\n",
      " |      True\n",
      " |      \n",
      " |      When we add the left and right SeqRecord objects, their annotation\n",
      " |      is all consistent, so it is all conserved in the new SeqRecord:\n",
      " |      \n",
      " |      >>> new.id == left.id == right.id == plasmid.id\n",
      " |      True\n",
      " |      >>> new.name == left.name == right.name == plasmid.name\n",
      " |      True\n",
      " |      >>> new.description == plasmid.description\n",
      " |      True\n",
      " |      >>> new.annotations == left.annotations == right.annotations\n",
      " |      True\n",
      " |      >>> new.letter_annotations == plasmid.letter_annotations\n",
      " |      True\n",
      " |      >>> new.dbxrefs == left.dbxrefs == right.dbxrefs\n",
      " |      True\n",
      " |      \n",
      " |      However, we should point out that when we sliced the SeqRecord,\n",
      " |      any annotations dictionary or dbxrefs list entries were lost.\n",
      " |      You can explicitly copy them like this:\n",
      " |      \n",
      " |      >>> new.annotations = plasmid.annotations.copy()\n",
      " |      >>> new.dbxrefs = plasmid.dbxrefs[:]\n",
      " |  \n",
      " |  __bool__(self)\n",
      " |      Boolean value of an instance of this class (True).\n",
      " |      \n",
      " |      This behaviour is for backwards compatibility, since until the\n",
      " |      __len__ method was added, a SeqRecord always evaluated as True.\n",
      " |      \n",
      " |      Note that in comparison, a Seq object will evaluate to False if it\n",
      " |      has a zero length sequence.\n",
      " |      \n",
      " |      WARNING: The SeqRecord may in future evaluate to False when its\n",
      " |      sequence is of zero length (in order to better match the Seq\n",
      " |      object behaviour)!\n",
      " |  \n",
      " |  __contains__(self, char)\n",
      " |      Implement the 'in' keyword, searches the sequence.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> \"GAATTC\" in record\n",
      " |      False\n",
      " |      >>> \"AAA\" in record\n",
      " |      True\n",
      " |      \n",
      " |      This essentially acts as a proxy for using \"in\" on the sequence:\n",
      " |      \n",
      " |      >>> \"GAATTC\" in record.seq\n",
      " |      False\n",
      " |      >>> \"AAA\" in record.seq\n",
      " |      True\n",
      " |      \n",
      " |      Note that you can also use Seq objects as the query,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> Seq(\"AAA\") in record\n",
      " |      True\n",
      " |      >>> Seq(\"AAA\", generic_dna) in record\n",
      " |      True\n",
      " |      \n",
      " |      See also the Seq object's __contains__ method.\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Define the equal-to operand (not implemented).\n",
      " |  \n",
      " |  __format__(self, format_spec)\n",
      " |      Return the record as a string in the specified file format.\n",
      " |      \n",
      " |      This method supports the Python format() function and f-strings.\n",
      " |      The format_spec should be a lower case string supported by\n",
      " |      Bio.SeqIO as a text output file format. Requesting a binary file\n",
      " |      format raises a ValueError. e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\"),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein\")\n",
      " |      ...\n",
      " |      >>> format(record, \"fasta\")\n",
      " |      '>YP_025292.1 toxic membrane protein\\nMKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\\n'\n",
      " |      >>> print(f\"Here is {record.id} in FASTA format:\\n{record:fasta}\")\n",
      " |      Here is YP_025292.1 in FASTA format:\n",
      " |      >YP_025292.1 toxic membrane protein\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      See also the SeqRecord's format() method.\n",
      " |  \n",
      " |  __ge__(self, other)\n",
      " |      Define the greater-than-or-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __getitem__(self, index)\n",
      " |      Return a sub-sequence or an individual letter.\n",
      " |      \n",
      " |      Slicing, e.g. my_record[5:10], returns a new SeqRecord for\n",
      " |      that sub-sequence with some annotation preserved as follows:\n",
      " |      \n",
      " |      * The name, id and description are kept as-is.\n",
      " |      * Any per-letter-annotations are sliced to match the requested\n",
      " |        sub-sequence.\n",
      " |      * Unless a stride is used, all those features which fall fully\n",
      " |        within the subsequence are included (with their locations\n",
      " |        adjusted accordingly). If you want to preserve any truncated\n",
      " |        features (e.g. GenBank/EMBL source features), you must\n",
      " |        explicitly add them to the new SeqRecord yourself.\n",
      " |      * The annotations dictionary and the dbxrefs list are not used\n",
      " |        for the new SeqRecord, as in general they may not apply to the\n",
      " |        subsequence. If you want to preserve them, you must explicitly\n",
      " |        copy them to the new SeqRecord yourself.\n",
      " |      \n",
      " |      Using an integer index, e.g. my_record[5] is shorthand for\n",
      " |      extracting that letter from the sequence, my_record.seq[5].\n",
      " |      \n",
      " |      For example, consider this short protein and its secondary\n",
      " |      structure as encoded by the PDB (e.g. H for alpha helices),\n",
      " |      plus a simple feature for its histidine self phosphorylation\n",
      " |      site:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> rec = SeqRecord(Seq(\"MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLAT\"\n",
      " |      ...                     \"EMMSEQDGYLAESINKDIEECNAIIEQFIDYLR\",\n",
      " |      ...                     IUPAC.protein),\n",
      " |      ...                 id=\"1JOY\", name=\"EnvZ\",\n",
      " |      ...                 description=\"Homodimeric domain of EnvZ from E. coli\")\n",
      " |      >>> rec.letter_annotations[\"secondary_structure\"] = \"  S  SSSSSSHHHHHTTTHHHHHHHHHHHHHHHHHHHHHHTHHHHHHHHHHHHHHHHHHHHHTT  \"\n",
      " |      >>> rec.features.append(SeqFeature(FeatureLocation(20, 21),\n",
      " |      ...                     type = \"Site\"))\n",
      " |      \n",
      " |      Now let's have a quick look at the full record,\n",
      " |      \n",
      " |      >>> print(rec)\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLATEMMSEQDGYLAESINKDIEE...YLR', IUPACProtein())\n",
      " |      >>> rec.letter_annotations[\"secondary_structure\"]\n",
      " |      '  S  SSSSSSHHHHHTTTHHHHHHHHHHHHHHHHHHHHHHTHHHHHHHHHHHHHHHHHHHHHTT  '\n",
      " |      >>> print(rec.features[0].location)\n",
      " |      [20:21]\n",
      " |      \n",
      " |      Now let's take a sub sequence, here chosen as the first (fractured)\n",
      " |      alpha helix which includes the histidine phosphorylation site:\n",
      " |      \n",
      " |      >>> sub = rec[11:41]\n",
      " |      >>> print(sub)\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('RTLLMAGVSHDLRTPLTRIRLATEMMSEQD', IUPACProtein())\n",
      " |      >>> sub.letter_annotations[\"secondary_structure\"]\n",
      " |      'HHHHHTTTHHHHHHHHHHHHHHHHHHHHHH'\n",
      " |      >>> print(sub.features[0].location)\n",
      " |      [9:10]\n",
      " |      \n",
      " |      You can also of course omit the start or end values, for\n",
      " |      example to get the first ten letters only:\n",
      " |      \n",
      " |      >>> print(rec[:10])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 0\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLAD', IUPACProtein())\n",
      " |      \n",
      " |      Or for the last ten letters:\n",
      " |      \n",
      " |      >>> print(rec[-10:])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 0\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('IIEQFIDYLR', IUPACProtein())\n",
      " |      \n",
      " |      If you omit both, then you get a copy of the original record (although\n",
      " |      lacking the annotations and dbxrefs):\n",
      " |      \n",
      " |      >>> print(rec[:])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLATEMMSEQDGYLAESINKDIEE...YLR', IUPACProtein())\n",
      " |      \n",
      " |      Finally, indexing with a simple integer is shorthand for pulling out\n",
      " |      that letter from the sequence directly:\n",
      " |      \n",
      " |      >>> rec[5]\n",
      " |      'K'\n",
      " |      >>> rec.seq[5]\n",
      " |      'K'\n",
      " |  \n",
      " |  __gt__(self, other)\n",
      " |      Define the greater-than operand (not implemented).\n",
      " |  \n",
      " |  __init__(self, seq, id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=None, features=None, annotations=None, letter_annotations=None)\n",
      " |      Create a SeqRecord.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - seq         - Sequence, required (Seq, MutableSeq or UnknownSeq)\n",
      " |       - id          - Sequence identifier, recommended (string)\n",
      " |       - name        - Sequence name, optional (string)\n",
      " |       - description - Sequence description, optional (string)\n",
      " |       - dbxrefs     - Database cross references, optional (list of strings)\n",
      " |       - features    - Any (sub)features, optional (list of SeqFeature objects)\n",
      " |       - annotations - Dictionary of annotations for the whole sequence\n",
      " |       - letter_annotations - Dictionary of per-letter-annotations, values\n",
      " |         should be strings, list or tuples of the same length as the full\n",
      " |         sequence.\n",
      " |      \n",
      " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |      SeqRecord objects.  However, you may want to create your own SeqRecord\n",
      " |      objects directly.\n",
      " |      \n",
      " |      Note that while an id is optional, we strongly recommend you supply a\n",
      " |      unique id string for each record.  This is especially important\n",
      " |      if you wish to write your sequences to a file.\n",
      " |      \n",
      " |      If you don't have the actual sequence, but you do know its length,\n",
      " |      then using the UnknownSeq object from Bio.Seq is appropriate.\n",
      " |      \n",
      " |      You can create a 'blank' SeqRecord object, and then populate the\n",
      " |      attributes later.\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |      Iterate over the letters in the sequence.\n",
      " |      \n",
      " |      For example, using Bio.SeqIO to read in a protein FASTA file:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/loveliesbleeding.pro\", \"fasta\")\n",
      " |      >>> for amino in record:\n",
      " |      ...     print(amino)\n",
      " |      ...     if amino == \"L\": break\n",
      " |      X\n",
      " |      A\n",
      " |      G\n",
      " |      L\n",
      " |      >>> print(record.seq[3])\n",
      " |      L\n",
      " |      \n",
      " |      This is just a shortcut for iterating over the sequence directly:\n",
      " |      \n",
      " |      >>> for amino in record.seq:\n",
      " |      ...     print(amino)\n",
      " |      ...     if amino == \"L\": break\n",
      " |      X\n",
      " |      A\n",
      " |      G\n",
      " |      L\n",
      " |      >>> print(record.seq[3])\n",
      " |      L\n",
      " |      \n",
      " |      Note that this does not facilitate iteration together with any\n",
      " |      per-letter-annotation.  However, you can achieve that using the\n",
      " |      python zip function on the record (or its sequence) and the relevant\n",
      " |      per-letter-annotation:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> rec = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (rec.id, rec.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(rec.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> for nuc, qual in zip(rec, rec.letter_annotations[\"solexa_quality\"]):\n",
      " |      ...     if qual > 35:\n",
      " |      ...         print(\"%s %i\" % (nuc, qual))\n",
      " |      A 40\n",
      " |      C 39\n",
      " |      G 38\n",
      " |      T 37\n",
      " |      A 36\n",
      " |      \n",
      " |      You may agree that using zip(rec.seq, ...) is more explicit than using\n",
      " |      zip(rec, ...) as shown above.\n",
      " |  \n",
      " |  __le___(self, other)\n",
      " |      Define the less-than-or-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the length of the sequence.\n",
      " |      \n",
      " |      For example, using Bio.SeqIO to read in a FASTA nucleotide file:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> len(record)\n",
      " |      309\n",
      " |      >>> len(record.seq)\n",
      " |      309\n",
      " |  \n",
      " |  __lt__(self, other)\n",
      " |      Define the less-than operand (not implemented).\n",
      " |  \n",
      " |  __ne__(self, other)\n",
      " |      Define the not-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add another sequence or string to this sequence (from the left).\n",
      " |      \n",
      " |      This method handles adding a Seq object (or similar, e.g. MutableSeq)\n",
      " |      or a plain Python string (on the left) to a SeqRecord (on the right).\n",
      " |      See the __add__ method for more details, but for example:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      \n",
      " |      >>> new = \"ACT\" + record\n",
      " |      >>> print(\"%s %s\" % (new.id, new.seq))\n",
      " |      slxa_0001_1_0001_01 ACTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(new.letter_annotations))\n",
      " |      []\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a concise summary of the record for debugging (string).\n",
      " |      \n",
      " |      The python built in function repr works by calling the object's ___repr__\n",
      " |      method.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Alphabet import generic_protein\n",
      " |      >>> rec = SeqRecord(Seq(\"MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKAT\"\n",
      " |      ...                    +\"GEMKEQTEWHRVVLFGKLAEVASEYLRKGSQVYIEGQLRTRKWTDQ\"\n",
      " |      ...                    +\"SGQDRYTTEVVVNVGGTMQMLGGRQGGGAPAGGNIGGGQPQGGWGQ\"\n",
      " |      ...                    +\"PQQPQGGNQFSGGAQSRPQQSAPAAPSNEPPMDFDDDIPF\",\n",
      " |      ...                    generic_protein),\n",
      " |      ...                 id=\"NP_418483.1\", name=\"b4059\",\n",
      " |      ...                 description=\"ssDNA-binding protein\",\n",
      " |      ...                 dbxrefs=[\"ASAP:13298\", \"GI:16131885\", \"GeneID:948570\"])\n",
      " |      >>> print(repr(rec))\n",
      " |      SeqRecord(seq=Seq('MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKATGEMKEQTE...IPF', ProteinAlphabet()), id='NP_418483.1', name='b4059', description='ssDNA-binding protein', dbxrefs=['ASAP:13298', 'GI:16131885', 'GeneID:948570'])\n",
      " |      \n",
      " |      At the python prompt you can also use this shorthand:\n",
      " |      \n",
      " |      >>> rec\n",
      " |      SeqRecord(seq=Seq('MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKATGEMKEQTE...IPF', ProteinAlphabet()), id='NP_418483.1', name='b4059', description='ssDNA-binding protein', dbxrefs=['ASAP:13298', 'GI:16131885', 'GeneID:948570'])\n",
      " |      \n",
      " |      Note that long sequences are shown truncated. Also note that any\n",
      " |      annotations, letter_annotations and features are not shown (as they\n",
      " |      would lead to a very long string).\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return a human readable summary of the record and its annotation (string).\n",
      " |      \n",
      " |      The python built in function str works by calling the object's ___str__\n",
      " |      method.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\",\n",
      " |      ...                         IUPAC.protein),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein, small\")\n",
      " |      >>> print(str(record))\n",
      " |      ID: YP_025292.1\n",
      " |      Name: HokC\n",
      " |      Description: toxic membrane protein, small\n",
      " |      Number of features: 0\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF', IUPACProtein())\n",
      " |      \n",
      " |      In this example you don't actually need to call str explicity, as the\n",
      " |      print command does this automatically:\n",
      " |      \n",
      " |      >>> print(record)\n",
      " |      ID: YP_025292.1\n",
      " |      Name: HokC\n",
      " |      Description: toxic membrane protein, small\n",
      " |      Number of features: 0\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF', IUPACProtein())\n",
      " |      \n",
      " |      Note that long sequences are shown truncated.\n",
      " |  \n",
      " |  format(self, format)\n",
      " |      Return the record as a string in the specified file format.\n",
      " |      \n",
      " |      The format should be a lower case string supported as an output\n",
      " |      format by Bio.SeqIO, which is used to turn the SeqRecord into a\n",
      " |      string.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\",\n",
      " |      ...                         IUPAC.protein),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein\")\n",
      " |      >>> record.format(\"fasta\")\n",
      " |      '>YP_025292.1 toxic membrane protein\\nMKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\\n'\n",
      " |      >>> print(record.format(\"fasta\"))\n",
      " |      >YP_025292.1 toxic membrane protein\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      The Python print function automatically appends a new line, meaning\n",
      " |      in this example a blank line is shown.  If you look at the string\n",
      " |      representation you can see there is a trailing new line (shown as\n",
      " |      slash n) which is important when writing to a file or if\n",
      " |      concatenating multiple sequence strings together.\n",
      " |      \n",
      " |      Note that this method will NOT work on every possible file format\n",
      " |      supported by Bio.SeqIO (e.g. some are for multiple sequences only,\n",
      " |      and binary formats are not supported).\n",
      " |  \n",
      " |  lower(self)\n",
      " |      Return a copy of the record with a lower case sequence.\n",
      " |      \n",
      " |      All the annotation is preserved unchanged. e.g.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/aster.pro\", \"fasta\")\n",
      " |      >>> print(record.format(\"fasta\"))\n",
      " |      >gi|3298468|dbj|BAA31520.1| SAMIPF\n",
      " |      GGHVNPAVTFGAFVGGNITLLRGIVYIIAQLLGSTVACLLLKFVTNDMAVGVFSLSAGVG\n",
      " |      VTNALVFEIVMTFGLVYTVYATAIDPKKGSLGTIAPIAIGFIVGANI\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(record.lower().format(\"fasta\"))\n",
      " |      >gi|3298468|dbj|BAA31520.1| SAMIPF\n",
      " |      gghvnpavtfgafvggnitllrgivyiiaqllgstvaclllkfvtndmavgvfslsagvg\n",
      " |      vtnalvfeivmtfglvytvyataidpkkgslgtiapiaigfivgani\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      To take a more annotation rich example,\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> old = SeqIO.read(\"EMBL/TRBG361.embl\", \"embl\")\n",
      " |      >>> len(old.features)\n",
      " |      3\n",
      " |      >>> new = old.lower()\n",
      " |      >>> len(old.features) == len(new.features)\n",
      " |      True\n",
      " |      >>> old.annotations[\"organism\"] == new.annotations[\"organism\"]\n",
      " |      True\n",
      " |      >>> old.dbxrefs == new.dbxrefs\n",
      " |      True\n",
      " |  \n",
      " |  reverse_complement(self, id=False, name=False, description=False, features=True, annotations=False, letter_annotations=True, dbxrefs=False)\n",
      " |      Return new SeqRecord with reverse complement sequence.\n",
      " |      \n",
      " |      By default the new record does NOT preserve the sequence identifier,\n",
      " |      name, description, general annotation or database cross-references -\n",
      " |      these are unlikely to apply to the reversed sequence.\n",
      " |      \n",
      " |      You can specify the returned record's id, name and description as\n",
      " |      strings, or True to keep that of the parent, or False for a default.\n",
      " |      \n",
      " |      You can specify the returned record's features with a list of\n",
      " |      SeqFeature objects, or True to keep that of the parent, or False to\n",
      " |      omit them. The default is to keep the original features (with the\n",
      " |      strand and locations adjusted).\n",
      " |      \n",
      " |      You can also specify both the returned record's annotations and\n",
      " |      letter_annotations as dictionaries, True to keep that of the parent,\n",
      " |      or False to omit them. The default is to keep the original\n",
      " |      annotations (with the letter annotations reversed).\n",
      " |      \n",
      " |      To show what happens to the pre-letter annotations, consider an\n",
      " |      example Solexa variant FASTQ file with a single entry, which we'll\n",
      " |      read in as a SeqRecord:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> print(record.letter_annotations[\"solexa_quality\"])\n",
      " |      [40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      Now take the reverse complement, here we explicitly give a new\n",
      " |      identifier (the old identifier with a suffix):\n",
      " |      \n",
      " |      >>> rc_record = record.reverse_complement(id=record.id + \"_rc\")\n",
      " |      >>> print(\"%s %s\" % (rc_record.id, rc_record.seq))\n",
      " |      slxa_0001_1_0001_01_rc NNNNNNACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT\n",
      " |      \n",
      " |      Notice that the per-letter-annotations have also been reversed,\n",
      " |      although this may not be appropriate for all cases.\n",
      " |      \n",
      " |      >>> print(rc_record.letter_annotations[\"solexa_quality\"])\n",
      " |      [-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40]\n",
      " |      \n",
      " |      Now for the features, we need a different example. Parsing a GenBank\n",
      " |      file is probably the easiest way to get an nice example with features\n",
      " |      in it...\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> with open(\"GenBank/pBAD30.gb\") as handle:\n",
      " |      ...     plasmid = SeqIO.read(handle, \"gb\")\n",
      " |      >>> print(\"%s %i\" % (plasmid.id, len(plasmid)))\n",
      " |      pBAD30 4923\n",
      " |      >>> plasmid.seq\n",
      " |      Seq('GCTAGCGGAGTGTATACTGGCTTACTATGTTGGCACTGATGAGGGTGTCAGTGA...ATG', IUPACAmbiguousDNA())\n",
      " |      >>> len(plasmid.features)\n",
      " |      13\n",
      " |      \n",
      " |      Now, let's take the reverse complement of this whole plasmid:\n",
      " |      \n",
      " |      >>> rc_plasmid = plasmid.reverse_complement(id=plasmid.id+\"_rc\")\n",
      " |      >>> print(\"%s %i\" % (rc_plasmid.id, len(rc_plasmid)))\n",
      " |      pBAD30_rc 4923\n",
      " |      >>> rc_plasmid.seq\n",
      " |      Seq('CATGGGCAAATATTATACGCAAGGCGACAAGGTGCTGATGCCGCTGGCGATTCA...AGC', IUPACAmbiguousDNA())\n",
      " |      >>> len(rc_plasmid.features)\n",
      " |      13\n",
      " |      \n",
      " |      Let's compare the first CDS feature - it has gone from being the\n",
      " |      second feature (index 1) to the second last feature (index -2), its\n",
      " |      strand has changed, and the location switched round.\n",
      " |      \n",
      " |      >>> print(plasmid.features[1])\n",
      " |      type: CDS\n",
      " |      location: [1081:1960](-)\n",
      " |      qualifiers:\n",
      " |          Key: label, Value: ['araC']\n",
      " |          Key: note, Value: ['araC regulator of the arabinose BAD promoter']\n",
      " |          Key: vntifkey, Value: ['4']\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(rc_plasmid.features[-2])\n",
      " |      type: CDS\n",
      " |      location: [2963:3842](+)\n",
      " |      qualifiers:\n",
      " |          Key: label, Value: ['araC']\n",
      " |          Key: note, Value: ['araC regulator of the arabinose BAD promoter']\n",
      " |          Key: vntifkey, Value: ['4']\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      You can check this new location, based on the length of the plasmid:\n",
      " |      \n",
      " |      >>> len(plasmid) - 1081\n",
      " |      3842\n",
      " |      >>> len(plasmid) - 1960\n",
      " |      2963\n",
      " |      \n",
      " |      Note that if the SeqFeature annotation includes any strand specific\n",
      " |      information (e.g. base changes for a SNP), this information is not\n",
      " |      amended, and would need correction after the reverse complement.\n",
      " |      \n",
      " |      Note trying to reverse complement a protein SeqRecord raises an\n",
      " |      exception:\n",
      " |      \n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> protein_rec = SeqRecord(Seq(\"MAIVMGR\", IUPAC.protein), id=\"Test\")\n",
      " |      >>> protein_rec.reverse_complement()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Proteins do not have complements!\n",
      " |      \n",
      " |      Also note you can reverse complement a SeqRecord using a MutableSeq:\n",
      " |      \n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Seq import MutableSeq\n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> rec = SeqRecord(MutableSeq(\"ACGT\", generic_dna), id=\"Test\")\n",
      " |      >>> rec.seq[0] = \"T\"\n",
      " |      >>> print(\"%s %s\" % (rec.id, rec.seq))\n",
      " |      Test TCGT\n",
      " |      >>> rc = rec.reverse_complement(id=True)\n",
      " |      >>> print(\"%s %s\" % (rc.id, rc.seq))\n",
      " |      Test ACGA\n",
      " |  \n",
      " |  translate(self, table='Standard', stop_symbol='*', to_stop=False, cds=False, gap=None, id=False, name=False, description=False, features=False, annotations=False, letter_annotations=False, dbxrefs=False)\n",
      " |      Return new SeqRecord with translated sequence.\n",
      " |      \n",
      " |      This calls the record's .seq.translate() method (which describes\n",
      " |      the translation related arguments, like table for the genetic code),\n",
      " |      \n",
      " |      By default the new record does NOT preserve the sequence identifier,\n",
      " |      name, description, general annotation or database cross-references -\n",
      " |      these are unlikely to apply to the translated sequence.\n",
      " |      \n",
      " |      You can specify the returned record's id, name and description as\n",
      " |      strings, or True to keep that of the parent, or False for a default.\n",
      " |      \n",
      " |      You can specify the returned record's features with a list of\n",
      " |      SeqFeature objects, or False (default) to omit them.\n",
      " |      \n",
      " |      You can also specify both the returned record's annotations and\n",
      " |      letter_annotations as dictionaries, True to keep that of the parent\n",
      " |      (annotations only), or False (default) to omit them.\n",
      " |      \n",
      " |      e.g. Loading a FASTA gene and translating it,\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> gene_record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> print(gene_record.format(\"fasta\"))\n",
      " |      >gi|3176602|gb|U78617.1|LOU78617 Lathyrus odoratus phytochrome A (PHYA) gene, partial cds\n",
      " |      CAGGCTGCGCGGTTTCTATTTATGAAGAACAAGGTCCGTATGATAGTTGATTGTCATGCA\n",
      " |      AAACATGTGAAGGTTCTTCAAGACGAAAAACTCCCATTTGATTTGACTCTGTGCGGTTCG\n",
      " |      ACCTTAAGAGCTCCACATAGTTGCCATTTGCAGTACATGGCTAACATGGATTCAATTGCT\n",
      " |      TCATTGGTTATGGCAGTGGTCGTCAATGACAGCGATGAAGATGGAGATAGCCGTGACGCA\n",
      " |      GTTCTACCACAAAAGAAAAAGAGACTTTGGGGTTTGGTAGTTTGTCATAACACTACTCCG\n",
      " |      AGGTTTGTT\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      And now translating the record, specifying the new ID and description:\n",
      " |      \n",
      " |      >>> protein_record = gene_record.translate(table=11,\n",
      " |      ...                                        id=\"phya\",\n",
      " |      ...                                        description=\"translation\")\n",
      " |      >>> print(protein_record.format(\"fasta\"))\n",
      " |      >phya translation\n",
      " |      QAARFLFMKNKVRMIVDCHAKHVKVLQDEKLPFDLTLCGSTLRAPHSCHLQYMANMDSIA\n",
      " |      SLVMAVVVNDSDEDGDSRDAVLPQKKKRLWGLVVCHNTTPRFV\n",
      " |      <BLANKLINE>\n",
      " |  \n",
      " |  upper(self)\n",
      " |      Return a copy of the record with an upper case sequence.\n",
      " |      \n",
      " |      All the annotation is preserved unchanged. e.g.\n",
      " |      \n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"acgtACGT\", generic_dna), id=\"Test\",\n",
      " |      ...                    description = \"Made up for this example\")\n",
      " |      >>> record.letter_annotations[\"phred_quality\"] = [1, 2, 3, 4, 5, 6, 7, 8]\n",
      " |      >>> print(record.upper().format(\"fastq\"))\n",
      " |      @Test Made up for this example\n",
      " |      ACGTACGT\n",
      " |      +\n",
      " |      \"#$%&'()\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      Naturally, there is a matching lower method:\n",
      " |      \n",
      " |      >>> print(record.lower().format(\"fastq\"))\n",
      " |      @Test Made up for this example\n",
      " |      acgtacgt\n",
      " |      +\n",
      " |      \"#$%&'()\n",
      " |      <BLANKLINE>\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  letter_annotations\n",
      " |      Dictionary of per-letter-annotation for the sequence.\n",
      " |      \n",
      " |      For example, this can hold quality scores used in FASTQ or QUAL files.\n",
      " |      Consider this example using Bio.SeqIO to read in an example Solexa\n",
      " |      variant FASTQ file as a SeqRecord:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> print(record.letter_annotations[\"solexa_quality\"])\n",
      " |      [40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      The letter_annotations get sliced automatically if you slice the\n",
      " |      parent SeqRecord, for example taking the last ten bases:\n",
      " |      \n",
      " |      >>> sub_record = record[-10:]\n",
      " |      >>> print(\"%s %s\" % (sub_record.id, sub_record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTNNNNNN\n",
      " |      >>> print(sub_record.letter_annotations[\"solexa_quality\"])\n",
      " |      [4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      Any python sequence (i.e. list, tuple or string) can be recorded in\n",
      " |      the SeqRecord's letter_annotations dictionary as long as the length\n",
      " |      matches that of the SeqRecord's sequence.  e.g.\n",
      " |      \n",
      " |      >>> len(sub_record.letter_annotations)\n",
      " |      1\n",
      " |      >>> sub_record.letter_annotations[\"dummy\"] = \"abcdefghij\"\n",
      " |      >>> len(sub_record.letter_annotations)\n",
      " |      2\n",
      " |      \n",
      " |      You can delete entries from the letter_annotations dictionary as usual:\n",
      " |      \n",
      " |      >>> del sub_record.letter_annotations[\"solexa_quality\"]\n",
      " |      >>> sub_record.letter_annotations\n",
      " |      {'dummy': 'abcdefghij'}\n",
      " |      \n",
      " |      You can completely clear the dictionary easily as follows:\n",
      " |      \n",
      " |      >>> sub_record.letter_annotations = {}\n",
      " |      >>> sub_record.letter_annotations\n",
      " |      {}\n",
      " |      \n",
      " |      Note that if replacing the record's sequence with a sequence of a\n",
      " |      different length you must first clear the letter_annotations dict.\n",
      " |  \n",
      " |  seq\n",
      " |      The sequence itself, as a Seq or MutableSeq object.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __hash__ = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.Alphabet import generic_dna\n",
    "fasta_file = SeqIO.parse(\"test.fa\",'fasta',alphabet=generic_dna) # Reconnect and parse the 'test.fa' file. This time we will specify an Alphabet for the records to indicate they are DNA sequences.\n",
    "\n",
    "# Lets grab the first record\n",
    "a = next(fasta_file)\n",
    "\n",
    "print(a)\n",
    "\n",
    "print(a.id)\n",
    "print(len(a))\n",
    "print(a.description)\n",
    "\n",
    "# To see a full list of available attributes (not all of them may be populated) we can use help()\n",
    "help(a)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And we can export or print the object in a variety of common biological formats (prespecified formats in [Bio.SeqIO](https://biopython.org/wiki/SeqIO). This makes for a very convenient way to convert sequence records between formats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dna_sequence_1\tTGGCGTTGTCTTTAATTCGATTAGCATTGCATATCGATTATCTAGCGATATGCTATGCTTAGC\n",
      "\n",
      "ID   dna_sequence_1; ; ; DNA; ; UNC; 63 BP.\n",
      "XX\n",
      "AC   dna_sequence_1;\n",
      "XX\n",
      "DE   dna_sequence_1\n",
      "XX\n",
      "OS   .\n",
      "OC   .\n",
      "XX\n",
      "FH   Key             Location/Qualifiers\n",
      "FH\n",
      "XX\n",
      "SQ   Sequence 63 BP; 14 A; 11 C; 13 G; 25 T; 0 other;\n",
      "     tggcgttgtc tttaattcga ttagcattgc atatcgatta tctagcgata tgctatgctt        60\n",
      "     agc                                                                      63\n",
      "//\n",
      "\n",
      "# STOCKHOLM 1.0\n",
      "#=GF SQ 1\n",
      "dna_sequence_1 TGGCGTTGTCTTTAATTCGATTAGCATTGCATATCGATTATCTAGCGATATGCTATGCTTAGC\n",
      "#=GS dna_sequence_1 AC dna_sequence_1\n",
      "#=GS dna_sequence_1 DE dna_sequence_1\n",
      "//\n",
      "\n",
      "CLUSTAL X (1.81) multiple sequence alignment\n",
      "\n",
      "\n",
      "dna_sequence_1                      TGGCGTTGTCTTTAATTCGATTAGCATTGCATATCGATTATCTAGCGATA\n",
      "\n",
      "dna_sequence_1                      TGCTATGCTTAGC\n",
      "\n",
      "\n",
      "\n",
      ">dna_sequence_1\n",
      "TGGCGTTGTCTTTAATTCGATTAGCATTGCATATCGATTATCTAGCGATATGCTATGCTT\n",
      "AGC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(a.format('tab'))\n",
    "print(a.format('embl'))\n",
    "print(a.format('stockholm'))\n",
    "print(a.format('clustal'))\n",
    "print(a.format('fasta'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can manipulate and slice `SeqRecord` objects as well. This will return a _new_ `SeqRecord` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('TGGCG', DNAAlphabet()), id='dna_sequence_1', name='dna_sequence_1', description='dna_sequence_1', dbxrefs=[])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a[:5] #retrieve the first 5 elements of the string"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can combine (concatenate) `SeqRecord` obejects together as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: dna_sequence_2\n",
      "Name: dna_sequence_2\n",
      "Description: dna_sequence_2\n",
      "Number of features: 0\n",
      "Seq('AGCCATATGTATCGAGCGATATGCGATACGAGTATCGAGTATGCAGTATGCATT...TGC', DNAAlphabet())\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('TGGCGTTGTCTTTAATTCGATTAGCATTGCATATCGATTATCTAGCGATATGCT...TGC', DNAAlphabet())\n",
      "126\n"
     ]
    }
   ],
   "source": [
    "b = next(fasta_file)\n",
    "\n",
    "print(b)\n",
    "\n",
    "new_seq = a + b\n",
    "\n",
    "print(new_seq)\n",
    "print(len(new_seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Querying NCBI datasets\n",
    "The National Center for Biotechnology Information (NCBI) maintains several useful data repositories and databases for biological data. [Entrez](https://www.ncbi.nlm.nih.gov/Web/Search/entrezfs.html) is a web-based data retrieval system that provides access to the NCBI databases. Biopython provides tools to access and retrieve data through Entrez from within your programs or scripts. Lets see some examples of how we can use these tools. First, lets get some information about what databases are available to query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b'<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\\n<!DOCTYPE eInfoResult PUBLIC \"-//NLM//DTD einfo 20190110//EN\" \"https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20190110/einfo.dtd\">\\n<eInfoResult>\\n<DbList>\\n\\n\\t<DbName>pubmed</DbName>\\n\\t<DbName>protein</DbName>\\n\\t<DbName>nuccore</DbName>\\n\\t<DbName>ipg</DbName>\\n\\t<DbName>nucleotide</DbName>\\n\\t<DbName>structure</DbName>\\n\\t<DbName>sparcle</DbName>\\n\\t<DbName>genome</DbName>\\n\\t<DbName>annotinfo</DbName>\\n\\t<DbName>assembly</DbName>\\n\\t<DbName>bioproject</DbName>\\n\\t<DbName>biosample</DbName>\\n\\t<DbName>blastdbinfo</DbName>\\n\\t<DbName>books</DbName>\\n\\t<DbName>cdd</DbName>\\n\\t<DbName>clinvar</DbName>\\n\\t<DbName>gap</DbName>\\n\\t<DbName>gapplus</DbName>\\n\\t<DbName>grasp</DbName>\\n\\t<DbName>dbvar</DbName>\\n\\t<DbName>gene</DbName>\\n\\t<DbName>gds</DbName>\\n\\t<DbName>geoprofiles</DbName>\\n\\t<DbName>homologene</DbName>\\n\\t<DbName>medgen</DbName>\\n\\t<DbName>mesh</DbName>\\n\\t<DbName>ncbisearch</DbName>\\n\\t<DbName>nlmcatalog</DbName>\\n\\t<DbName>omim</DbName>\\n\\t<DbName>orgtrack</DbName>\\n\\t<DbName>pmc</DbName>\\n\\t<DbName>popset</DbName>\\n\\t<DbName>proteinclusters</DbName>\\n\\t<DbName>pcassay</DbName>\\n\\t<DbName>biosystems</DbName>\\n\\t<DbName>pccompound</DbName>\\n\\t<DbName>pcsubstance</DbName>\\n\\t<DbName>seqannot</DbName>\\n\\t<DbName>snp</DbName>\\n\\t<DbName>sra</DbName>\\n\\t<DbName>taxonomy</DbName>\\n\\t<DbName>biocollections</DbName>\\n\\t<DbName>gtr</DbName>\\n</DbList>\\n\\n</eInfoResult>\\n'\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "# we must identify ourselves to the Entrez service in order to access\n",
    "Entrez.email = \"loyalgoff@jhmi.edu\"\n",
    "\n",
    "# make an initial connection in the form of a handle\n",
    "handle = Entrez.einfo()\n",
    "\n",
    "result = handle.read() # similar to reading a file\n",
    "\n",
    "handle.close() # close the handle when you are done\n",
    "\n",
    "print(result)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results are returned (by default) in XML format. But we can use Entrez's parser `Entrez.read` to read the handle and parse the XML directly into a more python-friendly dictionary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['DbList'])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make a new connection to Entrez\n",
    "handle = Entrez.einfo()\n",
    "\n",
    "result = Entrez.read(handle)\n",
    "\n",
    "result.keys() # returns a list of keys for the dictionary\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It seems this record has only one dictionary key called 'DbList' lets see whats inside"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'sparcle', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'homologene', 'medgen', 'mesh', 'ncbisearch', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'popset', 'proteinclusters', 'pcassay', 'biosystems', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result['DbList']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a list of all of the available databases which we can query with Entrez using the Biopython interface. We can recursively learn more about each of these databases by usine EInfo again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['DbName', 'MenuName', 'Description', 'DbBuild', 'Count', 'LastUpdate', 'FieldList', 'LinkList'])\n",
      "Core Nucleotide db\n",
      "2020/08/24 17:07\n"
     ]
    }
   ],
   "source": [
    "handle = Entrez.einfo(db=\"nucleotide\")\n",
    "\n",
    "result = Entrez.read(handle)\n",
    "\n",
    "print(result[\"DbInfo\"].keys())\n",
    "\n",
    "print(result[\"DbInfo\"][\"Description\"])\n",
    "print(result[\"DbInfo\"][\"LastUpdate\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `FieldList` element is a list of possible search fields for this database. Let's dig a bit deeper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ALL\t All Fields\t All terms from all searchable fields\n",
      "UID\t UID\t Unique number assigned to each sequence\n",
      "FILT\t Filter\t Limits the records\n",
      "WORD\t Text Word\t Free text associated with record\n",
      "TITL\t Title\t Words in definition line\n",
      "KYWD\t Keyword\t Nonstandardized terms provided by submitter\n",
      "AUTH\t Author\t Author(s) of publication\n",
      "JOUR\t Journal\t Journal abbreviation of publication\n",
      "VOL\t Volume\t Volume number of publication\n",
      "ISS\t Issue\t Issue number of publication\n",
      "PAGE\t Page Number\t Page number(s) of publication\n",
      "ORGN\t Organism\t Scientific and common names of organism, and all higher levels of taxonomy\n",
      "ACCN\t Accession\t Accession number of sequence\n",
      "PACC\t Primary Accession\t Does not include retired secondary accessions\n",
      "GENE\t Gene Name\t Name of gene associated with sequence\n",
      "PROT\t Protein Name\t Name of protein associated with sequence\n",
      "ECNO\t EC/RN Number\t EC number for enzyme or CAS registry number\n",
      "PDAT\t Publication Date\t Date sequence added to GenBank\n",
      "MDAT\t Modification Date\t Date of last update\n",
      "SUBS\t Substance Name\t CAS chemical name or MEDLINE Substance Name\n",
      "PROP\t Properties\t Classification by source qualifiers and molecule type\n",
      "SQID\t SeqID String\t String identifier for sequence\n",
      "GPRJ\t BioProject\t BioProject\n",
      "SLEN\t Sequence Length\t Length of sequence\n",
      "FKEY\t Feature key\t Feature annotated on sequence\n",
      "PORG\t Primary Organism\t Scientific and common names of primary organism, and all higher levels of taxonomy\n",
      "COMP\t Component Accession\t Component accessions for an assembly\n",
      "ASSM\t Assembly\t Assembly\n",
      "DIV\t Division\t Division\n",
      "STRN\t Strain\t Strain\n",
      "ISOL\t Isolate\t Isolate\n",
      "CULT\t Cultivar\t Cultivar\n",
      "BRD\t Breed\t Breed\n",
      "BIOS\t BioSample\t BioSample\n"
     ]
    }
   ],
   "source": [
    "for field in result['DbInfo']['FieldList']:\n",
    "    print(f'{field[\"Name\"]}\\t {field[\"FullName\"]}\\t {field[\"Description\"]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now you try and get information about another NCBI database through `Entrez.einfo()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Searching for gene information and ids\n",
    "Now we can use this information to begin to query the nucleotide database at NCBI using the above fields to construct our search. Lets see if we can search records associated with the human gene 'Notch1' using `esearch()` "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Count': '24', 'RetMax': '20', 'RetStart': '0', 'IdList': ['1389542951', '1034665363', '568815589', '170014708', '1147710141', '150249057', '74273673', '71516264', '584614703', '444738984', '452887541', '452887540', '452887539', '452887538', '452887537', '452887536', '452887535', '452887534', '380509012', '379331273'], 'TranslationSet': [{'From': 'Homo[Orgn]', 'To': '\"Homo\"[Organism]'}], 'TranslationStack': [{'Term': '\"Homo\"[Organism]', 'Field': 'Organism', 'Count': '27675084', 'Explode': 'Y'}, {'Term': 'NOTCH1[Gene]', 'Field': 'Gene', 'Count': '1087', 'Explode': 'N'}, 'AND'], 'QueryTranslation': '\"Homo\"[Organism] AND NOTCH1[Gene]'}\n",
      "24\n"
     ]
    }
   ],
   "source": [
    "handle = Entrez.esearch(\n",
    "                db = 'nucleotide',\n",
    "                term = \"Homo[Orgn] AND NOTCH1[Gene]\"\n",
    "    )\n",
    "\n",
    "result = Entrez.read(handle)\n",
    "\n",
    "print(result)\n",
    "\n",
    "print(result['Count'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have identified 24 records associated with this gene.  Here we can see the Entrez IDs assocated with each record.  We will use these to 'fetch' the full record for the first few id records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['1389542951', '1034665363', '568815589', '170014708', '1147710141', '150249057', '74273673', '71516264', '584614703', '444738984', '452887541', '452887540', '452887539', '452887538', '452887537', '452887536', '452887535', '452887534', '380509012', '379331273']\n"
     ]
    }
   ],
   "source": [
    "notch1_ids = result['IdList']\n",
    "print(str(notch1_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NM_017617.5 Homo sapiens notch receptor 1 (NOTCH1), mRNA\n",
      "ACTAGTGCCTCGGCCGCGGGAGGGAGCGCAAGGGCGCGGGGCGCGGGGCGCGGGCGCGGG\n",
      "CGCGAGCGCAGCGAAGGAACGAGCCGGGCGCGGAGCCGGGCCCGGGGGCCTGCGAGAGCA\n",
      "CAGCGCCGCCAGCCAGCCGGGGAAGAGAGGGCGGGACCGTCCGCCGCCGCCCCGGGACCG\n",
      "TACGCCGCGCGTGTGCGTCCCAGCCCCGCCGGCCAGCGCAGGAGGCCGCCGCCCGGGCGC\n",
      "AGAGGGCAGCCGGTGGGGAGGCATGCCGCCGCTCCTGGCGCCCCTGCTCTGCCTGGCGCT\n",
      "GCTGCCCGCGCTCGCCGCACGAGGCCCGCGATGCTCCCAGCCCGGTGAGACCTGCCTGAA\n",
      "TGGCGGGAAGTGTGAAGCGGCCAATGGCACGGAGGCCTGCGTCTGTGGCGGGGCCTTCGT\n",
      "GGGCCCGCGATGCCAGGACCCCAACCCGTGCCTCAGCACCCCCTGCAAGAACGCCGGGAC\n",
      "ATGCCACGTGGTGGACCGCAGAGGCGTGGCAGACTATGCCTGCAGCTGTGCCCTGGGCTT\n",
      "CTCTGGGCCCCTCTGCCTGACACCCCTGGACAATGCCTGCCTCACCAACCCCTGCCGCAA\n",
      "CGGGGGCACCTGCGACCTGCTCACGCTGACGGAGTACAAGTGCCGCTGCCCGCCCGGCTG\n",
      "GTCAGGGAAATCGTGCCAGCAGGCTGACCCGTGCGCCTCCAACCCCTGCGCCAACGGTGG\n",
      "CCAGTGCCTGCCCTTCGAGGCCTCCTACATCTGCCACTGCCCACCCAGCTTCCATGGCCC\n",
      "CACCTGCCGGCAGGATGTCAACGAGTGTGGCCAGAAGCCCGGGCTTTGCCGCCACGGAGG\n",
      "CACCTGCCACAACGAGGTCGGCTCCTACCGCTGCGTCTGCCGCGCCACCCACACTGGCCC\n",
      "CAACTGCGAGCGGCCCTACGTGCCCTGCAGCCCCTCGCCCTGCCAGAACGGGGGCACCTG\n",
      "CCGCCCCACGGGCGACGTCACCCACGAGTGTGCCTGCCTGCCAGGCTTCACCGGCCAGAA\n",
      "CTGTGAGGAAAATATCGACGATTGTCCAGGAAACAACTGCAAGAACGGGGGTGCCTGTGT\n",
      "GGACGGCGTGAACACCTACAACTGCCGCTGCCCGCCAGAGTGGACAGGTCAGTACTGTAC\n",
      "CGAGGATGTGGACGAGTGCCAGCTGATGCCAAATGCCTGCCAGAACGGCGGGACCTGCCA\n",
      "CAACACCCACGGTGGCTACAACTGCGTGTGTGTCAACGGCTGGACTGGTGAGGACTGCAG\n",
      "CGAGAACATTGATGACTGTGCCAGCGCCGCCTGCTTCCACGGCGCCACCTGCCATGACCG\n",
      "TGTGGCCTCCTTCTACTGCGAGTGTCCCCATGGCCGCACAGGTCTGCTGTGCCACCTCAA\n",
      "CGACGCATGCATCAGCAACCCCTGTAACGAGGGCTCCAACTGCGACACCAACCCTGTCAA\n",
      "TGGCAAGGCCATCTGCACCTGCCCCTCGGGGTACACGGGCCCGGCCTGCAGCCAGGACGT\n",
      "GGATGAGTGCTCGCTGGGTGCCAACCCCTGCGAGCATGCGGGCAAGTGCATCAACACGCT\n",
      "GGGCTCCTTCGAGTGCCAGTGTCTGCAGGGCTACACGGGCCCCCGATGCGAGATCGACGT\n",
      "CAACGAGTGCGTCTCGAACCCGTGCCAGAACGACGCCACCTGCCTGGACCAGATTGGGGA\n",
      "GTTCCAGTGCATCTGCATGCCCGGCTACGAGGGTGTGCACTGCGAGGTCAACACAGACGA\n",
      "GTGTGCCAGCAGCCCCTGCCTGCACAATGGCCGCTGCCTGGACAAGATCAATGAGTTCCA\n",
      "GTGCGAGTGCCCCACGGGCTTCACTGGGCATCTGTGCCAGTACGATGTGGACGAGTGTGC\n",
      "CAGCACCCCCTGCAAGAATGGTGCCAAGTGCCTGGACGGACCCAACACTTACACCTGTGT\n",
      "GTGCACGGAAGGGTACACGGGGACGCACTGCGAGGTGGACATCGATGAGTGCGACCCCGA\n",
      "CCCCTGCCACTACGGCTCCTGCAAGGACGGCGTCGCCACCTTCACCTGCCTCTGCCGCCC\n",
      "AGGCTACACGGGCCACCACTGCGAGACCAACATCAACGAGTGCTCCAGCCAGCCCTGCCG\n",
      "CCACGGGGGCACCTGCCAGGACCGCGACAACGCCTACCTCTGCTTCTGCCTGAAGGGGAC\n",
      "CACAGGACCCAACTGCGAGATCAACCTGGATGACTGTGCCAGCAGCCCCTGCGACTCGGG\n",
      "CACCTGTCTGGACAAGATCGATGGCTACGAGTGTGCCTGTGAGCCGGGCTACACAGGGAG\n",
      "CATGTGTAACATCAACATCGATGAGTGTGCGGGCAACCCCTGCCACAACGGGGGCACCTG\n",
      "CGAGGACGGCATCAATGGCTTCACCTGCCGCTGCCCCGAGGGCTACCACGACCCCACCTG\n",
      "CCTGTCTGAGGTCAATGAGTGCAACAGCAACCCCTGCGTCCACGGGGCCTGCCGGGACAG\n",
      "CCTCAACGGGTACAAGTGCGACTGTGACCCTGGGTGGAGTGGGACCAACTGTGACATCAA\n",
      "CAACAATGAGTGTGAATCCAACCCTTGTGTCAACGGCGGCACCTGCAAAGACATGACCAG\n",
      "TGGCTACGTGTGCACCTGCCGGGAGGGCTTCAGCGGTCCCAACTGCCAGACCAACATCAA\n",
      "CGAGTGTGCGTCCAACCCATGTCTGAACCAGGGCACGTGTATTGACGACGTTGCCGGGTA\n",
      "CAAGTGCAACTGCCTGCTGCCCTACACAGGTGCCACGTGTGAGGTGGTGCTGGCCCCGTG\n",
      "TGCCCCCAGCCCCTGCAGAAACGGCGGGGAGTGCAGGCAATCCGAGGACTATGAGAGCTT\n",
      "CTCCTGTGTCTGCCCCACGGGCTGGCAAGGGCAGACCTGTGAGGTCGACATCAACGAGTG\n",
      "CGTTCTGAGCCCGTGCCGGCACGGCGCATCCTGCCAGAACACCCACGGCGGCTACCGCTG\n",
      "CCACTGCCAGGCCGGCTACAGTGGGCGCAACTGCGAGACCGACATCGACGACTGCCGGCC\n",
      "CAACCCGTGTCACAACGGGGGCTCCTGCACAGACGGCATCAACACGGCCTTCTGCGACTG\n",
      "CCTGCCCGGCTTCCGGGGCACTTTCTGTGAGGAGGACATCAACGAGTGTGCCAGTGACCC\n",
      "CTGCCGCAACGGGGCCAACTGCACGGACTGCGTGGACAGCTACACGTGCACCTGCCCCGC\n",
      "AGGCTTCAGCGGGATCCACTGTGAGAACAACACGCCTGACTGCACAGAGAGCTCCTGCTT\n",
      "CAACGGTGGCACCTGCGTGGACGGCATCAACTCGTTCACCTGCCTGTGTCCACCCGGCTT\n",
      "CACGGGCAGCTACTGCCAGCACGATGTCAATGAGTGCGACTCACAGCCCTGCCTGCATGG\n",
      "CGGCACCTGTCAGGACGGCTGCGGCTCCTACAGGTGCACCTGCCCCCAGGGCTACACTGG\n",
      "CCCCAACTGCCAGAACCTTGTGCACTGGTGTGACTCCTCGCCCTGCAAGAACGGCGGCAA\n",
      "ATGCTGGCAGACCCACACCCAGTACCGCTGCGAGTGCCCCAGCGGCTGGACCGGCCTTTA\n",
      "CTGCGACGTGCCCAGCGTGTCCTGTGAGGTGGCTGCGCAGCGACAAGGTGTTGACGTTGC\n",
      "CCGCCTGTGCCAGCATGGAGGGCTCTGTGTGGACGCGGGCAACACGCACCACTGCCGCTG\n",
      "CCAGGCGGGCTACACAGGCAGCTACTGTGAGGACCTGGTGGACGAGTGCTCACCCAGCCC\n",
      "CTGCCAGAACGGGGCCACCTGCACGGACTACCTGGGCGGCTACTCCTGCAAGTGCGTGGC\n",
      "CGGCTACCACGGGGTGAACTGCTCTGAGGAGATCGACGAGTGCCTCTCCCACCCCTGCCA\n",
      "GAACGGGGGCACCTGCCTCGACCTCCCCAACACCTACAAGTGCTCCTGCCCACGGGGCAC\n",
      "TCAGGGTGTGCACTGTGAGATCAACGTGGACGACTGCAATCCCCCCGTTGACCCCGTGTC\n",
      "CCGGAGCCCCAAGTGCTTTAACAACGGCACCTGCGTGGACCAGGTGGGCGGCTACAGCTG\n",
      "CACCTGCCCGCCGGGCTTCGTGGGTGAGCGCTGTGAGGGGGATGTCAACGAGTGCCTGTC\n",
      "CAATCCCTGCGACGCCCGTGGCACCCAGAACTGCGTGCAGCGCGTCAATGACTTCCACTG\n",
      "CGAGTGCCGTGCTGGTCACACCGGGCGCCGCTGCGAGTCCGTCATCAATGGCTGCAAAGG\n",
      "CAAGCCCTGCAAGAATGGGGGCACCTGCGCCGTGGCCTCCAACACCGCCCGCGGGTTCAT\n",
      "CTGCAAGTGCCCTGCGGGCTTCGAGGGCGCCACGTGTGAGAATGACGCTCGTACCTGCGG\n",
      "CAGCCTGCGCTGCCTCAACGGCGGCACATGCATCTCCGGCCCGCGCAGCCCCACCTGCCT\n",
      "GTGCCTGGGCCCCTTCACGGGCCCCGAATGCCAGTTCCCGGCCAGCAGCCCCTGCCTGGG\n",
      "CGGCAACCCCTGCTACAACCAGGGGACCTGTGAGCCCACATCCGAGAGCCCCTTCTACCG\n",
      "TTGCCTGTGCCCCGCCAAATTCAACGGGCTCTTGTGCCACATCCTGGACTACAGCTTCGG\n",
      "GGGTGGGGCCGGGCGCGACATCCCCCCGCCGCTGATCGAGGAGGCGTGCGAGCTGCCCGA\n",
      "GTGCCAGGAGGACGCGGGCAACAAGGTCTGCAGCCTGCAGTGCAACAACCACGCGTGCGG\n",
      "CTGGGACGGCGGTGACTGCTCCCTCAACTTCAATGACCCCTGGAAGAACTGCACGCAGTC\n",
      "TCTGCAGTGCTGGAAGTACTTCAGTGACGGCCACTGTGACAGCCAGTGCAACTCAGCCGG\n",
      "CTGCCTCTTCGACGGCTTTGACTGCCAGCGTGCGGAAGGCCAGTGCAACCCCCTGTACGA\n",
      "CCAGTACTGCAAGGACCACTTCAGCGACGGGCACTGCGACCAGGGCTGCAACAGCGCGGA\n",
      "GTGCGAGTGGGACGGGCTGGACTGTGCGGAGCATGTACCCGAGAGGCTGGCGGCCGGCAC\n",
      "GCTGGTGGTGGTGGTGCTGATGCCGCCGGAGCAGCTGCGCAACAGCTCCTTCCACTTCCT\n",
      "GCGGGAGCTCAGCCGCGTGCTGCACACCAACGTGGTCTTCAAGCGTGACGCACACGGCCA\n",
      "GCAGATGATCTTCCCCTACTACGGCCGCGAGGAGGAGCTGCGCAAGCACCCCATCAAGCG\n",
      "TGCCGCCGAGGGCTGGGCCGCACCTGACGCCCTGCTGGGCCAGGTGAAGGCCTCGCTGCT\n",
      "CCCTGGTGGCAGCGAGGGTGGGCGGCGGCGGAGGGAGCTGGACCCCATGGACGTCCGCGG\n",
      "CTCCATCGTCTACCTGGAGATTGACAACCGGCAGTGTGTGCAGGCCTCCTCGCAGTGCTT\n",
      "CCAGAGTGCCACCGACGTGGCCGCATTCCTGGGAGCGCTCGCCTCGCTGGGCAGCCTCAA\n",
      "CATCCCCTACAAGATCGAGGCCGTGCAGAGTGAGACCGTGGAGCCGCCCCCGCCGGCGCA\n",
      "GCTGCACTTCATGTACGTGGCGGCGGCCGCCTTTGTGCTTCTGTTCTTCGTGGGCTGCGG\n",
      "GGTGCTGCTGTCCCGCAAGCGCCGGCGGCAGCATGGCCAGCTCTGGTTCCCTGAGGGCTT\n",
      "CAAAGTGTCTGAGGCCAGCAAGAAGAAGCGGCGGGAGCCCCTCGGCGAGGACTCCGTGGG\n",
      "CCTCAAGCCCCTGAAGAACGCTTCAGACGGTGCCCTCATGGACGACAACCAGAATGAGTG\n",
      "GGGGGACGAGGACCTGGAGACCAAGAAGTTCCGGTTCGAGGAGCCCGTGGTTCTGCCTGA\n",
      "CCTGGACGACCAGACAGACCACCGGCAGTGGACTCAGCAGCACCTGGATGCCGCTGACCT\n",
      "GCGCATGTCTGCCATGGCCCCCACACCGCCCCAGGGTGAGGTTGACGCCGACTGCATGGA\n",
      "CGTCAATGTCCGCGGGCCTGATGGCTTCACCCCGCTCATGATCGCCTCCTGCAGCGGGGG\n",
      "CGGCCTGGAGACGGGCAACAGCGAGGAAGAGGAGGACGCGCCGGCCGTCATCTCCGACTT\n",
      "CATCTACCAGGGCGCCAGCCTGCACAACCAGACAGACCGCACGGGCGAGACCGCCTTGCA\n",
      "CCTGGCCGCCCGCTACTCACGCTCTGATGCCGCCAAGCGCCTGCTGGAGGCCAGCGCAGA\n",
      "TGCCAACATCCAGGACAACATGGGCCGCACCCCGCTGCATGCGGCTGTGTCTGCCGACGC\n",
      "ACAAGGTGTCTTCCAGATCCTGATCCGGAACCGAGCCACAGACCTGGATGCCCGCATGCA\n",
      "TGATGGCACGACGCCACTGATCCTGGCTGCCCGCCTGGCCGTGGAGGGCATGCTGGAGGA\n",
      "CCTCATCAACTCACACGCCGACGTCAACGCCGTAGATGACCTGGGCAAGTCCGCCCTGCA\n",
      "CTGGGCCGCCGCCGTGAACAATGTGGATGCCGCAGTTGTGCTCCTGAAGAACGGGGCTAA\n",
      "CAAAGATATGCAGAACAACAGGGAGGAGACACCCCTGTTTCTGGCCGCCCGGGAGGGCAG\n",
      "CTACGAGACCGCCAAGGTGCTGCTGGACCACTTTGCCAACCGGGACATCACGGATCATAT\n",
      "GGACCGCCTGCCGCGCGACATCGCACAGGAGCGCATGCATCACGACATCGTGAGGCTGCT\n",
      "GGACGAGTACAACCTGGTGCGCAGCCCGCAGCTGCACGGAGCCCCGCTGGGGGGCACGCC\n",
      "CACCCTGTCGCCCCCGCTCTGCTCGCCCAACGGCTACCTGGGCAGCCTCAAGCCCGGCGT\n",
      "GCAGGGCAAGAAGGTCCGCAAGCCCAGCAGCAAAGGCCTGGCCTGTGGAAGCAAGGAGGC\n",
      "CAAGGACCTCAAGGCACGGAGGAAGAAGTCCCAGGACGGCAAGGGCTGCCTGCTGGACAG\n",
      "CTCCGGCATGCTCTCGCCCGTGGACTCCCTGGAGTCACCCCATGGCTACCTGTCAGACGT\n",
      "GGCCTCGCCGCCACTGCTGCCCTCCCCGTTCCAGCAGTCTCCGTCCGTGCCCCTCAACCA\n",
      "CCTGCCTGGGATGCCCGACACCCACCTGGGCATCGGGCACCTGAACGTGGCGGCCAAGCC\n",
      "CGAGATGGCGGCGCTGGGTGGGGGCGGCCGGCTGGCCTTTGAGACTGGCCCACCTCGTCT\n",
      "CTCCCACCTGCCTGTGGCCTCTGGCACCAGCACCGTCCTGGGCTCCAGCAGCGGAGGGGC\n",
      "CCTGAATTTCACTGTGGGCGGGTCCACCAGTTTGAATGGTCAATGCGAGTGGCTGTCCCG\n",
      "GCTGCAGAGCGGCATGGTGCCGAACCAATACAACCCTCTGCGGGGGAGTGTGGCACCAGG\n",
      "CCCCCTGAGCACACAGGCCCCCTCCCTGCAGCATGGCATGGTAGGCCCGCTGCACAGTAG\n",
      "CCTTGCTGCCAGCGCCCTGTCCCAGATGATGAGCTACCAGGGCCTGCCCAGCACCCGGCT\n",
      "GGCCACCCAGCCTCACCTGGTGCAGACCCAGCAGGTGCAGCCACAAAACTTACAGATGCA\n",
      "GCAGCAGAACCTGCAGCCAGCAAACATCCAGCAGCAGCAAAGCCTGCAGCCGCCACCACC\n",
      "ACCACCACAGCCGCACCTTGGCGTGAGCTCAGCAGCCAGCGGCCACCTGGGCCGGAGCTT\n",
      "CCTGAGTGGAGAGCCGAGCCAGGCAGACGTGCAGCCACTGGGCCCCAGCAGCCTGGCGGT\n",
      "GCACACTATTCTGCCCCAGGAGAGCCCCGCCCTGCCCACGTCGCTGCCATCCTCGCTGGT\n",
      "CCCACCCGTGACCGCAGCCCAGTTCCTGACGCCCCCCTCGCAGCACAGCTACTCCTCGCC\n",
      "TGTGGACAACACCCCCAGCCACCAGCTACAGGTGCCTGAGCACCCCTTCCTCACCCCGTC\n",
      "CCCTGAGTCCCCTGACCAGTGGTCCAGCTCGTCCCCGCATTCCAACGTCTCCGACTGGTC\n",
      "CGAGGGCGTCTCCAGCCCTCCCACCAGCATGCAGTCCCAGATCGCCCGCATTCCGGAGGC\n",
      "CTTCAAGTAAACGGCGCGCCCCACGAGACCCCGGCTTCCTTTCCCAAGCCTTCGGGCGTC\n",
      "TGTGTGCGCTCTGTGGATGCCAGGGCCGACCAGAGGAGCCTTTTTAAAACACATGTTTTT\n",
      "ATACAAAATAAGAACGAGGATTTTAATTTTTTTTAGTATTTATTTATGTACTTTTATTTT\n",
      "ACACAGAAACACTGCCTTTTTATTTATATGTACTGTTTTATCTGGCCCCAGGTAGAAACT\n",
      "TTTATCTATTCTGAGAAAACAAGCAAGTTCTGAGAGCCAGGGTTTTCCTACGTAGGATGA\n",
      "AAAGATTCTTCTGTGTTTATAAAATATAAACAAAGATTCATGATTTATAAATGCCATTTA\n",
      "TTTATTGATTCCTTTTTTCAAAATCCAAAAAGAAATGATGTTGGAGAAGGGAAGTTGAAC\n",
      "GAGCATAGTCCAAAAAGCTCCTGGGGCGTCCAGGCCGCGCCCTTTCCCCGACGCCCACCC\n",
      "AACCCCAAGCCAGCCCGGCCGCTCCACCAGCATCACCTGCCTGTTAGGAGAAGCTGCATC\n",
      "CAGAGGCAAACGGAGGCAAAGCTGGCTCACCTTCCGCACGCGGATTAATTTGCATCTGAA\n",
      "ATAGGAAACAAGTGAAAGCATATGGGTTAGATGTTGCCATGTGTTTTAGATGGTTTCTTG\n",
      "CAAGCATGCTTGTGAAAATGTGTTCTCGGAGTGTGTATGCCAAGAGTGCACCCATGGTAC\n",
      "CAATCATGAATCTTTGTTTCAGGTTCAGTATTATGTAGTTGTTCGTTGGTTATACAAGTT\n",
      "CTTGGTCCCTCCAGAACCACCCCGGCCCCCTGCCCGTTCTTGAAATGTAGGCATCATGCA\n",
      "TGTCAAACATGAGATGTGTGGACTGTGGCACTTGCCTGGGTCACACACGGAGGCATCCTA\n",
      "CCCTTTTCTGGGGAAAGACACTGCCTGGGCTGACCCCGGTGGCGGCCCCAGCACCTCAGC\n",
      "CTGCACAGTGTCCCCCAGGTTCCGAAGAAGATGCTCCAGCAACACAGCCTGGGCCCCAGC\n",
      "TCGCGGGACCCGACCCCCCGTGGGCTCCCGTGTTTTGTAGGAGACTTGCCAGAGCCGGGC\n",
      "ACATTGAGCTGTGCAACGCCGTGGGCTGCGTCCTTTGGTCCTGTCCCCGCAGCCCTGGCA\n",
      "GGGGGCATGCGGTCGGGCAGGGGCTGGAGGGAGGCGGGGGCTGCCCTTGGGCCACCCCTC\n",
      "CTAGTTTGGGAGGAGCAGATTTTTGCAATACCAAGTATAGCCTATGGCAGAAAAAATGTC\n",
      "TGTAAATATGTTTTTAAAGGTGGATTTTGTTTAAAAAATCTTAATGAATGAGTCTGTTGT\n",
      "GTGTCATGCCAGTGAGGGACGTCAGACTTGGCTCAGCTCGGGGAGCCTTAGCCGCCCATG\n",
      "CACTGGGGACGCTCCGCTGCCGTGCCGCCTGCACTCCTCAGGGCAGCCTCCCCCGGCTCT\n",
      "ACGGGGGCCGCGTGGTGCCATCCCCAGGGGGCATGACCAGATGCGTCCCAAGATGTTGAT\n",
      "TTTTACTGTGTTTTATAAAATAGAGTGTAGTTTACAGAAAAAGACTTTAAAAGTGATCTA\n",
      "CATGAGGAACTGTAGATGATGTATTTTTTTCATCTTTTTTGTTAACTGATTTGCAATAAA\n",
      "AATGATACTGATGGTGATCTGGCTTCCA\n",
      "\n",
      ">XM_011518717.2 PREDICTED: Homo sapiens notch receptor 1 (NOTCH1), transcript variant X1, mRNA\n",
      "AGTGTGAAGCGGCCAATGGCACGGAGGCCTGCGTGCTTCACCGGCCAGAACTGTGAGGAA\n",
      "AATATCGACGATTGTCCAGGAAACAACTGCAAGAACGGGGGTGCCTGTGTGGACGGCGTG\n",
      "AACACCTACAACTGCCGCTGCCCGCCAGAGTGGACAGGTCAGTACTGTACCGAGGATGTG\n",
      "GACGAGTGCCAGCTGATGCCAAATGCCTGCCAGAACGGCGGGACCTGCCACAACACCCAC\n",
      "GGTGGCTACAACTGCGTGTGTGTCAACGGCTGGACTGGTGAGGACTGCAGCGAGAACATT\n",
      "GATGACTGTGCCAGCGCCGCCTGCTTCCACGGCGCCACCTGCCATGACCGTGTGGCCTCC\n",
      "TTCTACTGCGAGTGTCCCCATGGCCGCACAGGTCTGCTGTGCCACCTCAACGACGCATGC\n",
      "ATCAGCAACCCCTGTAACGAGGGCTCCAACTGCGACACCAACCCTGTCAATGGCAAGGCC\n",
      "ATCTGCACCTGCCCCTCGGGGTACACGGGCCCGGCCTGCAGCCAGGACGTGGATGAGTGC\n",
      "TCGCTGGGTGCCAACCCCTGCGAGCATGCGGGCAAGTGCATCAACACGCTGGGCTCCTTC\n",
      "GAGTGCCAGTGTCTGCAGGGCTACACGGGCCCCCGATGCGAGATCGACGTCAACGAGTGC\n",
      "GTCTCGAACCCGTGCCAGAACGACGCCACCTGCCTGGACCAGATTGGGGAGTTCCAGTGC\n",
      "ATCTGCATGCCCGGCTACGAGGGTGTGCACTGCGAGGTCAACACAGACGAGTGTGCCAGC\n",
      "AGCCCCTGCCTGCACAATGGCCGCTGCCTGGACAAGATCAATGAGTTCCAGTGCGAGTGC\n",
      "CCCACGGGCTTCACTGGGCATCTGTGCCAGTACGATGTGGACGAGTGTGCCAGCACCCCC\n",
      "TGCAAGAATGGTGCCAAGTGCCTGGACGGACCCAACACTTACACCTGTGTGTGCACGGAA\n",
      "GGGTACACGGGGACGCACTGCGAGGTGGACATCGATGAGTGCGACCCCGACCCCTGCCAC\n",
      "TACGGCTCCTGCAAGGACGGCGTCGCCACCTTCACCTGCCTCTGCCGCCCAGGCTACACG\n",
      "GGCCACCACTGCGAGACCAACATCAACGAGTGCTCCAGCCAGCCCTGCCGCCACGGGGGC\n",
      "ACCTGCCAGGACCGCGACAACGCCTACCTCTGCTTCTGCCTGAAGGGGACCACAGGACCC\n",
      "AACTGCGAGATCAACCTGGATGACTGTGCCAGCAGCCCCTGCGACTCGGGCACCTGTCTG\n",
      "GACAAGATCGATGGCTACGAGTGTGCCTGTGAGCCGGGCTACACAGGGAGCATGTGTAAC\n",
      "ATCAACATCGATGAGTGTGCGGGCAACCCCTGCCACAACGGGGGCACCTGCGAGGACGGC\n",
      "ATCAATGGCTTCACCTGCCGCTGCCCCGAGGGCTACCACGACCCCACCTGCCTGTCTGAG\n",
      "GTCAATGAGTGCAACAGCAACCCCTGCGTCCACGGGGCCTGCCGGGACAGCCTCAACGGG\n",
      "TACAAGTGCGACTGTGACCCTGGGTGGAGTGGGACCAACTGTGACATCAACAACAATGAG\n",
      "TGTGAATCCAACCCTTGTGTCAACGGCGGCACCTGCAAAGACATGACCAGTGGCTACGTG\n",
      "TGCACCTGCCGGGAGGGCTTCAGCGGTCCCAACTGCCAGACCAACATCAACGAGTGTGCG\n",
      "TCCAACCCATGTCTGAACCAGGGCACGTGTATTGACGACGTTGCCGGGTACAAGTGCAAC\n",
      "TGCCTGCTGCCCTACACAGGTGCCACGTGTGAGGTGGTGCTGGCCCCGTGTGCCCCCAGC\n",
      "CCCTGCAGAAACGGCGGGGAGTGCAGGCAATCCGAGGACTATGAGAGCTTCTCCTGTGTC\n",
      "TGCCCCACGGGCTGGCAAGGGCAGACCTGTGAGGTCGACATCAACGAGTGCGTTCTGAGC\n",
      "CCGTGCCGGCACGGCGCATCCTGCCAGAACACCCACGGCGGCTACCGCTGCCACTGCCAG\n",
      "GCCGGCTACAGTGGGCGCAACTGCGAGACCGACATCGACGACTGCCGGCCCAACCCGTGT\n",
      "CACAACGGGGGCTCCTGCACAGACGGCATCAACACGGCCTTCTGCGACTGCCTGCCCGGC\n",
      "TTCCGGGGCACTTTCTGTGAGGAGGACATCAACGAGTGTGCCAGTGACCCCTGCCGCAAC\n",
      "GGGGCCAACTGCACGGACTGCGTGGACAGCTACACGTGCACCTGCCCCGCAGGCTTCAGC\n",
      "GGGATCCACTGTGAGAACAACACGCCTGACTGCACAGAGAGCTCCTGCTTCAACGGTGGC\n",
      "ACCTGCGTGGACGGCATCAACTCGTTCACCTGCCTGTGTCCACCCGGCTTCACGGGCAGC\n",
      "TACTGCCAGCACGATGTCAATGAGTGCGACTCACAGCCCTGCCTGCATGGCGGCACCTGT\n",
      "CAGGACGGCTGCGGCTCCTACAGGTGCACCTGCCCCCAGGGCTACACTGGCCCCAACTGC\n",
      "CAGAACCTTGTGCACTGGTGTGACTCCTCGCCCTGCAAGAACGGCGGCAAATGCTGGCAG\n",
      "ACCCACACCCAGTACCGCTGCGAGTGCCCCAGCGGCTGGACCGGCCTTTACTGCGACGTG\n",
      "CCCAGCGTGTCCTGTGAGGTGGCTGCGCAGCGACAAGGTGTTGACGTTGCCCGCCTGTGC\n",
      "CAGCATGGAGGGCTCTGTGTGGACGCGGGCAACACGCACCACTGCCGCTGCCAGGCGGGC\n",
      "TACACAGGCAGCTACTGTGAGGACCTGGTGGACGAGTGCTCACCCAGCCCCTGCCAGAAC\n",
      "GGGGCCACCTGCACGGACTACCTGGGCGGCTACTCCTGCAAGTGCGTGGCCGGCTACCAC\n",
      "GGGGTGAACTGCTCTGAGGAGATCGACGAGTGCCTCTCCCACCCCTGCCAGAACGGGGGC\n",
      "ACCTGCCTCGACCTCCCCAACACCTACAAGTGCTCCTGCCCACGGGGCACTCAGGGTGTG\n",
      "CACTGTGAGATCAACGTGGACGACTGCAATCCCCCCGTTGACCCCGTGTCCCGGAGCCCC\n",
      "AAGTGCTTTAACAACGGCACCTGCGTGGACCAGGTGGGCGGCTACAGCTGCACCTGCCCG\n",
      "CCGGGCTTCGTGGGTGAGCGCTGTGAGGGGGATGTCAACGAGTGCCTGTCCAATCCCTGC\n",
      "GACGCCCGTGGCACCCAGAACTGCGTGCAGCGCGTCAATGACTTCCACTGCGAGTGCCGT\n",
      "GCTGGTCACACCGGGCGCCGCTGCGAGTCCGTCATCAATGGCTGCAAAGGCAAGCCCTGC\n",
      "AAGAATGGGGGCACCTGCGCCGTGGCCTCCAACACCGCCCGCGGGTTCATCTGCAAGTGC\n",
      "CCTGCGGGCTTCGAGGGCGCCACGTGTGAGAATGACGCTCGTACCTGCGGCAGCCTGCGC\n",
      "TGCCTCAACGGCGGCACATGCATCTCCGGCCCGCGCAGCCCCACCTGCCTGTGCCTGGGC\n",
      "CCCTTCACGGGCCCCGAATGCCAGTTCCCGGCCAGCAGCCCCTGCCTGGGCGGCAACCCC\n",
      "TGCTACAACCAGGGGACCTGTGAGCCCACATCCGAGAGCCCCTTCTACCGTTGCCTGTGC\n",
      "CCCGCCAAATTCAACGGGCTCTTGTGCCACATCCTGGACTACAGCTTCGGGGGTGGGGCC\n",
      "GGGCGCGACATCCCCCCGCCGCTGATCGAGGAGGCGTGCGAGCTGCCCGAGTGCCAGGAG\n",
      "GACGCGGGCAACAAGGTCTGCAGCCTGCAGTGCAACAACCACGCGTGCGGCTGGGACGGC\n",
      "GGTGACTGCTCCCTCAACTTCAATGACCCCTGGAAGAACTGCACGCAGTCTCTGCAGTGC\n",
      "TGGAAGTACTTCAGTGACGGCCACTGTGACAGCCAGTGCAACTCAGCCGGCTGCCTCTTC\n",
      "GACGGCTTTGACTGCCAGCGTGCGGAAGGCCAGTGCAACCCCCTGTACGACCAGTACTGC\n",
      "AAGGACCACTTCAGCGACGGGCACTGCGACCAGGGCTGCAACAGCGCGGAGTGCGAGTGG\n",
      "GACGGGCTGGACTGTGCGGAGCATGTACCCGAGAGGCTGGCGGCCGGCACGCTGGTGGTG\n",
      "GTGGTGCTGATGCCGCCGGAGCAGCTGCGCAACAGCTCCTTCCACTTCCTGCGGGAGCTC\n",
      "AGCCGCGTGCTGCACACCAACGTGGTCTTCAAGCGTGACGCACACGGCCAGCAGATGATC\n",
      "TTCCCCTACTACGGCCGCGAGGAGGAGCTGCGCAAGCACCCCATCAAGCGTGCCGCCGAG\n",
      "GGCTGGGCCGCACCTGACGCCCTGCTGGGCCAGGTGAAGGCCTCGCTGCTCCCTGGTGGC\n",
      "AGCGAGGGTGGGCGGCGGCGGAGGGAGCTGGACCCCATGGACGTCCGCGGCTCCATCGTC\n",
      "TACCTGGAGATTGACAACCGGCAGTGTGTGCAGGCCTCCTCGCAGTGCTTCCAGAGTGCC\n",
      "ACCGACGTGGCCGCATTCCTGGGAGCGCTCGCCTCGCTGGGCAGCCTCAACATCCCCTAC\n",
      "AAGATCGAGGCCGTGCAGAGTGAGACCGTGGAGCCGCCCCCGCCGGCGCAGCTGCACTTC\n",
      "ATGTACGTGGCGGCGGCCGCCTTTGTGCTTCTGTTCTTCGTGGGCTGCGGGGTGCTGCTG\n",
      "TCCCGCAAGCGCCGGCGGCAGCATGGCCAGCTCTGGTTCCCTGAGGGCTTCAAAGTGTCT\n",
      "GAGGCCAGCAAGAAGAAGCGGCGGGAGCCCCTCGGCGAGGACTCCGTGGGCCTCAAGCCC\n",
      "CTGAAGAACGCTTCAGACGGTGCCCTCATGGACGACAACCAGAATGAGTGGGGGGACGAG\n",
      "GACCTGGAGACCAAGAAGTTCCGGTTCGAGGAGCCCGTGGTTCTGCCTGACCTGGACGAC\n",
      "CAGACAGACCACCGGCAGTGGACTCAGCAGCACCTGGATGCCGCTGACCTGCGCATGTCT\n",
      "GCCATGGCCCCCACACCGCCCCAGGGTGAGGTTGACGCCGACTGCATGGACGTCAATGTC\n",
      "CGCGGGCCTGATGGCTTCACCCCGCTCATGATCGCCTCCTGCAGCGGGGGCGGCCTGGAG\n",
      "ACGGGCAACAGCGAGGAAGAGGAGGACGCGCCGGCCGTCATCTCCGACTTCATCTACCAG\n",
      "GGCGCCAGCCTGCACAACCAGACAGACCGCACGGGCGAGACCGCCTTGCACCTGGCCGCC\n",
      "CGCTACTCACGCTCTGATGCCGCCAAGCGCCTGCTGGAGGCCAGCGCAGATGCCAACATC\n",
      "CAGGACAACATGGGCCGCACCCCGCTGCATGCGGCTGTGTCTGCCGACGCACAAGGTGTC\n",
      "TTCCAGATCCTGATCCGGAACCGAGCCACAGACCTGGATGCCCGCATGCATGATGGCACG\n",
      "ACGCCACTGATCCTGGCTGCCCGCCTGGCCGTGGAGGGCATGCTGGAGGACCTCATCAAC\n",
      "TCACACGCCGACGTCAACGCCGTAGATGACCTGGGCAAGTCCGCCCTGCACTGGGCCGCC\n",
      "GCCGTGAACAATGTGGATGCCGCAGTTGTGCTCCTGAAGAACGGGGCTAACAAAGATATG\n",
      "CAGAACAACAGGGAGGAGACACCCCTGTTTCTGGCCGCCCGGGAGGGCAGCTACGAGACC\n",
      "GCCAAGGTGCTGCTGGACCACTTTGCCAACCGGGACATCACGGATCATATGGACCGCCTG\n",
      "CCGCGCGACATCGCACAGGAGCGCATGCATCACGACATCGTGAGGCTGCTGGACGAGTAC\n",
      "AACCTGGTGCGCAGCCCGCAGCTGCACGGAGCCCCGCTGGGGGGCACGCCCACCCTGTCG\n",
      "CCCCCGCTCTGCTCGCCCAACGGCTACCTGGGCAGCCTCAAGCCCGGCGTGCAGGGCAAG\n",
      "AAGGTCCGCAAGCCCAGCAGCAAAGGCCTGGCCTGTGGAAGCAAGGAGGCCAAGGACCTC\n",
      "AAGGCACGGAGGAAGAAGTCCCAGGACGGCAAGGGCTGCCTGCTGGACAGCTCCGGCATG\n",
      "CTCTCGCCCGTGGACTCCCTGGAGTCACCCCATGGCTACCTGTCAGACGTGGCCTCGCCG\n",
      "CCACTGCTGCCCTCCCCGTTCCAGCAGTCTCCGTCCGTGCCCCTCAACCACCTGCCTGGG\n",
      "ATGCCCGACACCCACCTGGGCATCGGGCACCTGAACGTGGCGGCCAAGCCCGAGATGGCG\n",
      "GCGCTGGGTGGGGGCGGCCGGCTGGCCTTTGAGACTGGCCCACCTCGTCTCTCCCACCTG\n",
      "CCTGTGGCCTCTGGCACCAGCACCGTCCTGGGCTCCAGCAGCGGAGGGGCCCTGAATTTC\n",
      "ACTGTGGGCGGGTCCACCAGTTTGAATGGTCAATGCGAGTGGCTGTCCCGGCTGCAGAGC\n",
      "GGCATGGTGCCGAACCAATACAACCCTCTGCGGGGGAGTGTGGCACCAGGCCCCCTGAGC\n",
      "ACACAGGCCCCCTCCCTGCAGCATGGCATGGTAGGCCCGCTGCACAGTAGCCTTGCTGCC\n",
      "AGCGCCCTGTCCCAGATGATGAGCTACCAGGGCCTGCCCAGCACCCGGCTGGCCACCCAG\n",
      "CCTCACCTGGTGCAGACCCAGCAGGTGCAGCCACAAAACTTACAGATGCAGCAGCAGAAC\n",
      "CTGCAGCCAGCAAACATCCAGCAGCAGCAAAGCCTGCAGCCGCCACCACCACCACCACAG\n",
      "CCGCACCTTGGCGTGAGCTCAGCAGCCAGCGGCCACCTGGGCCGGAGCTTCCTGAGTGGA\n",
      "GAGCCGAGCCAGGCAGACGTGCAGCCACTGGGCCCCAGCAGCCTGGCGGTGCACACTATT\n",
      "CTGCCCCAGGAGAGCCCCGCCCTGCCCACGTCGCTGCCATCCTCGCTGGTCCCACCCGTG\n",
      "ACCGCAGCCCAGTTCCTGACGCCCCCCTCGCAGCACAGCTACTCCTCGCCTGTGGACAAC\n",
      "ACCCCCAGCCACCAGCTACAGGTGCCTGAGCACCCCTTCCTCACCCCGTCCCCTGAGTCC\n",
      "CCTGACCAGTGGTCCAGCTCGTCCCCGCATTCCAACGTCTCCGACTGGTCCGAGGGCGTC\n",
      "TCCAGCCCTCCCACCAGCATGCAGTCCCAGATCGCCCGCATTCCGGAGGCCTTCAAGTAA\n",
      "ACGGCGCGCCCCACGAGACCCCGGCTTCCTTTCCCAAGCCTTCGGGCGTCTGTGTGCGCT\n",
      "CTGTGGATGCCAGGGCCGACCAGAGGAGCCTTTTTAAAACACATGTTTTTATACAAAATA\n",
      "AGAACGAGGATTTTAATTTTTTTTAGTATTTATTTATGTACTTTTATTTTACACAGAAAC\n",
      "ACTGCCTTTTTATTTATATGTACTGTTTTATCTGGCCCCAGGTAGAAACTTTTATCTATT\n",
      "CTGAGAAAACAAGCAAGTTCTGAGAGCCAGGGTTTTCCTACGTAGGATGAAAAGATTCTT\n",
      "CTGTGTTTATAAAATATAAACAAAGATTCATGATTTATAAATGCCATTTATTTATTGATT\n",
      "CCTTTTTTCAAAATCCAAAAAGAAATGATGTTGGAGAAGGGAAGTTGAACGAGCATAGTC\n",
      "CAAAAAGCTCCTGGGGCGTCCAGGCCGCGCCCTTTCCCCGACGCCCACCCAACCCCAAGC\n",
      "CAGCCCGGCCGCTCCACCAGCATCACCTGCCTGTTAGGAGAAGCTGCATCCAGAGGCAAA\n",
      "CGGAGGCAAAGCTGGCTCACCTTCCGCACGCGGATTAATTTGCATCTGAAATAGGAAACA\n",
      "AGTGAAAGCATATGGGTTAGATGTTGCCATGTGTTTTAGATGGTTTCTTGCAAGCATGCT\n",
      "TGTGAAAATGTGTTCTCGGAGTGTGTATGCCAAGAGTGCACCCATGGTACCAATCATGAA\n",
      "TCTTTGTTTCAGGTTCAGTATTATGTAGTTGTTCGTTGGTTATACAAGTTCTTGGTCCCT\n",
      "CCAGAACCACCCCGGCCCCCTGCCCGTTCTTGAAATGTAGGCATCATGCATGTCAAACAT\n",
      "GAGATGTGTGGACTGTGGCACTTGCCTGGGTCACACACGGAGGCATCCTACCCTTTTCTG\n",
      "GGGAAAGACACTGCCTGGGCTGACCCCGGTGGCGGCCCCAGCACCTCAGCCTGCACAGTG\n",
      "TCCCCCAGGTTCCGAAGAAGATGCTCCAGCAACACAGCCTGGGCCCCAGCTCGCGGGACC\n",
      "CGACCCCCCGTGGGCTCCCGTGTTTTGTAGGAGACTTGCCAGAGCCGGGCACATTGAGCT\n",
      "GTGCAACGCCGTGGGCTGCGTCCTTTGGTCCTGTCCCCGCAGCCCTGGCAGGGGGCATGC\n",
      "GGTCGGGCAGGGGCTGGAGGGAGGCGGGGGCTGCCCTTGGGCCACCCCTCCTAGTTTGGG\n",
      "AGGAGCAGATTTTTGCAATACCAAGTATAGCCTATGGCAGAAAAAATGTCTGTAAATATG\n",
      "TTTTTAAAGGTGGATTTTGTTTAAAAAATCTTAA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "handle = Entrez.efetch(db=\"nucleotide\", id=notch1_ids[:2], rettype=\"fasta\", retmode=\"text\") # here we are specifically requesting these records in a fasta format\n",
    "\n",
    "results = SeqIO.parse(handle,'fasta') # and now we can pass this handle to Bio.SeqIO.parse to parse the handle into SeqRecord objects\n",
    "\n",
    "# and now we can loop through the list of 'SeqRecord' objects\n",
    "for res in results:\n",
    "    print(res.format('fasta'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Of course there are a number of different 'pythonic' ways in which you can achieve this as well.  For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene ID NM_017617.5 contains 107 features\n"
     ]
    }
   ],
   "source": [
    "with Entrez.efetch(\n",
    "    db=\"nucleotide\", rettype=\"gb\", retmode=\"text\", id=\"1389542951\"\n",
    ") as handle:\n",
    "    record = SeqIO.read(handle, \"gb\")\n",
    "print(f'Gene ID {record.id} contains {len(record.features)} features')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type: source\n",
      "location: [0:9568](+)\n",
      "qualifiers:\n",
      "    Key: chromosome, Value: ['9']\n",
      "    Key: db_xref, Value: ['taxon:9606']\n",
      "    Key: map, Value: ['9q34.3']\n",
      "    Key: mol_type, Value: ['mRNA']\n",
      "    Key: organism, Value: ['Homo sapiens']\n",
      "\n",
      "type: gene\n",
      "location: [0:9568](+)\n",
      "qualifiers:\n",
      "    Key: db_xref, Value: ['GeneID:4851', 'HGNC:HGNC:7881', 'MIM:190198']\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['notch receptor 1']\n",
      "\n",
      "type: exon\n",
      "location: [0:323](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: inference, Value: ['alignment:Splign:2.1.0']\n",
      "\n",
      "type: CDS\n",
      "location: [262:7930](+)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['CCDS:CCDS43905.1', 'GeneID:4851', 'HGNC:HGNC:7881', 'MIM:190198']\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['translocation-associated notch protein TAN-1; neurogenic locus notch homolog protein 1; Notch homolog 1, translocation-associated; notch 1']\n",
      "    Key: product, Value: ['neurogenic locus notch homolog protein 1 preproprotein']\n",
      "    Key: protein_id, Value: ['NP_060087.3']\n",
      "    Key: translation, Value: ['MPPLLAPLLCLALLPALAARGPRCSQPGETCLNGGKCEAANGTEACVCGGAFVGPRCQDPNPCLSTPCKNAGTCHVVDRRGVADYACSCALGFSGPLCLTPLDNACLTNPCRNGGTCDLLTLTEYKCRCPPGWSGKSCQQADPCASNPCANGGQCLPFEASYICHCPPSFHGPTCRQDVNECGQKPGLCRHGGTCHNEVGSYRCVCRATHTGPNCERPYVPCSPSPCQNGGTCRPTGDVTHECACLPGFTGQNCEENIDDCPGNNCKNGGACVDGVNTYNCRCPPEWTGQYCTEDVDECQLMPNACQNGGTCHNTHGGYNCVCVNGWTGEDCSENIDDCASAACFHGATCHDRVASFYCECPHGRTGLLCHLNDACISNPCNEGSNCDTNPVNGKAICTCPSGYTGPACSQDVDECSLGANPCEHAGKCINTLGSFECQCLQGYTGPRCEIDVNECVSNPCQNDATCLDQIGEFQCICMPGYEGVHCEVNTDECASSPCLHNGRCLDKINEFQCECPTGFTGHLCQYDVDECASTPCKNGAKCLDGPNTYTCVCTEGYTGTHCEVDIDECDPDPCHYGSCKDGVATFTCLCRPGYTGHHCETNINECSSQPCRHGGTCQDRDNAYLCFCLKGTTGPNCEINLDDCASSPCDSGTCLDKIDGYECACEPGYTGSMCNINIDECAGNPCHNGGTCEDGINGFTCRCPEGYHDPTCLSEVNECNSNPCVHGACRDSLNGYKCDCDPGWSGTNCDINNNECESNPCVNGGTCKDMTSGYVCTCREGFSGPNCQTNINECASNPCLNQGTCIDDVAGYKCNCLLPYTGATCEVVLAPCAPSPCRNGGECRQSEDYESFSCVCPTGWQGQTCEVDINECVLSPCRHGASCQNTHGGYRCHCQAGYSGRNCETDIDDCRPNPCHNGGSCTDGINTAFCDCLPGFRGTFCEEDINECASDPCRNGANCTDCVDSYTCTCPAGFSGIHCENNTPDCTESSCFNGGTCVDGINSFTCLCPPGFTGSYCQHDVNECDSQPCLHGGTCQDGCGSYRCTCPQGYTGPNCQNLVHWCDSSPCKNGGKCWQTHTQYRCECPSGWTGLYCDVPSVSCEVAAQRQGVDVARLCQHGGLCVDAGNTHHCRCQAGYTGSYCEDLVDECSPSPCQNGATCTDYLGGYSCKCVAGYHGVNCSEEIDECLSHPCQNGGTCLDLPNTYKCSCPRGTQGVHCEINVDDCNPPVDPVSRSPKCFNNGTCVDQVGGYSCTCPPGFVGERCEGDVNECLSNPCDARGTQNCVQRVNDFHCECRAGHTGRRCESVINGCKGKPCKNGGTCAVASNTARGFICKCPAGFEGATCENDARTCGSLRCLNGGTCISGPRSPTCLCLGPFTGPECQFPASSPCLGGNPCYNQGTCEPTSESPFYRCLCPAKFNGLLCHILDYSFGGGAGRDIPPPLIEEACELPECQEDAGNKVCSLQCNNHACGWDGGDCSLNFNDPWKNCTQSLQCWKYFSDGHCDSQCNSAGCLFDGFDCQRAEGQCNPLYDQYCKDHFSDGHCDQGCNSAECEWDGLDCAEHVPERLAAGTLVVVVLMPPEQLRNSSFHFLRELSRVLHTNVVFKRDAHGQQMIFPYYGREEELRKHPIKRAAEGWAAPDALLGQVKASLLPGGSEGGRRRRELDPMDVRGSIVYLEIDNRQCVQASSQCFQSATDVAAFLGALASLGSLNIPYKIEAVQSETVEPPPPAQLHFMYVAAAAFVLLFFVGCGVLLSRKRRRQHGQLWFPEGFKVSEASKKKRREPLGEDSVGLKPLKNASDGALMDDNQNEWGDEDLETKKFRFEEPVVLPDLDDQTDHRQWTQQHLDAADLRMSAMAPTPPQGEVDADCMDVNVRGPDGFTPLMIASCSGGGLETGNSEEEEDAPAVISDFIYQGASLHNQTDRTGETALHLAARYSRSDAAKRLLEASADANIQDNMGRTPLHAAVSADAQGVFQILIRNRATDLDARMHDGTTPLILAARLAVEGMLEDLINSHADVNAVDDLGKSALHWAAAVNNVDAAVVLLKNGANKDMQNNREETPLFLAAREGSYETAKVLLDHFANRDITDHMDRLPRDIAQERMHHDIVRLLDEYNLVRSPQLHGAPLGGTPTLSPPLCSPNGYLGSLKPGVQGKKVRKPSSKGLACGSKEAKDLKARRKKSQDGKGCLLDSSGMLSPVDSLESPHGYLSDVASPPLLPSPFQQSPSVPLNHLPGMPDTHLGIGHLNVAAKPEMAALGGGGRLAFETGPPRLSHLPVASGTSTVLGSSSGGALNFTVGGSTSLNGQCEWLSRLQSGMVPNQYNPLRGSVAPGPLSTQAPSLQHGMVGPLHSSLAASALSQMMSYQGLPSTRLATQPHLVQTQQVQPQNLQMQQQNLQPANIQQQQSLQPPPPPPQPHLGVSSAASGHLGRSFLSGEPSQADVQPLGPSSLAVHTILPQESPALPTSLPSSLVPPVTAAQFLTPPSQHSYSSPVDNTPSHQLQVPEHPFLTPSPESPDQWSSSSPHSNVSDWSEGVSSPPTSMQSQIARIPEAFK']\n",
      "\n",
      "type: sig_peptide\n",
      "location: [262:316](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: inference, Value: ['COORDINATES: ab initio prediction:SignalP:4.0']\n",
      "\n",
      "type: mat_peptide\n",
      "location: [316:7927](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['propagated from UniProtKB/Swiss-Prot (P46531.4)']\n",
      "    Key: product, Value: ['Neurogenic locus notch homolog protein 1. /id=PRO_0000007674']\n",
      "\n",
      "type: misc_feature\n",
      "location: [382:385](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['N-linked (GlcNAc...) asparagine. /evidence=ECO:0000255; propagated from UniProtKB/Swiss-Prot (P46531.4); glycosylation site']\n",
      "\n",
      "type: misc_feature\n",
      "location: [454:457](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['O-linked (Glc...) serine. /evidence=ECO:0000250|UniProtKB:Q01705; propagated from UniProtKB/Swiss-Prot (P46531.4); glycosylation site']\n",
      "\n",
      "type: misc_feature\n",
      "location: [478:481](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['O-linked (Fuc...) threonine. /evidence=ECO:0000250|UniProtKB:Q01705; propagated from UniProtKB/Swiss-Prot (P46531.4); glycosylation site']\n",
      "\n",
      "type: misc_feature\n",
      "location: [607:610](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['O-linked (Fuc...) threonine. /evidence=ECO:0000250|UniProtKB:Q01705; propagated from UniProtKB/Swiss-Prot (P46531.4); glycosylation site']\n",
      "\n",
      "type: misc_feature\n",
      "location: [697:700](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['O-linked (Glc...) serine. /evidence=ECO:0000250|UniProtKB:Q01705; propagated from UniProtKB/Swiss-Prot (P46531.4); glycosylation site']\n",
      "\n",
      "type: misc_feature\n",
      "location: [841:844](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['O-linked (Fuc...) threonine. /evidence=ECO:0000250|UniProtKB:Q01705; propagated from UniProtKB/Swiss-Prot (P46531.4); glycosylation site']\n",
      "\n",
      "type: misc_feature\n",
      "location: [955:958](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['O-linked (Fuc...) threonine, alternate. /evidence=ECO:0000269|PubMed:24226769; O-linked (GalNAc...) threonine, alternate. /evidence=ECO:0000269|PubMed:24226769; propagated from UniProtKB/Swiss-Prot (P46531.4); glycosylation site']\n",
      "\n",
      "type: misc_feature\n",
      "location: [1192:1195](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['O-linked (Fuc...) threonine. /evidence=ECO:0000250|UniProtKB:Q01705; propagated from UniProtKB/Swiss-Prot (P46531.4); glycosylation site']\n",
      "\n",
      "type: misc_feature\n",
      "location: [1282:1285](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['O-linked (Glc...) serine. /evidence=ECO:0000250|UniProtKB:Q01705; propagated from UniProtKB/Swiss-Prot (P46531.4); glycosylation site']\n",
      "\n",
      "type: misc_feature\n",
      "location: [1306:1309](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['O-linked (Fuc...) threonine. /evidence=ECO:0000250|UniProtKB:Q01705; propagated from UniProtKB/Swiss-Prot (P46531.4); glycosylation site']\n",
      "\n",
      "type: misc_feature\n",
      "location: [1393:1396](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['O-linked (Glc...) serine. /evidence=ECO:0000250|UniProtKB:Q01705; propagated from UniProtKB/Swiss-Prot (P46531.4); glycosylation site']\n",
      "\n",
      "type: misc_feature\n",
      "location: [1519:1525](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['propagated from UniProtKB/Swiss-Prot (P46531.4); Region: Interaction with DLL4. /evidence=ECO:0000250|UniProtKB:Q07008']\n",
      "\n",
      "type: misc_feature\n",
      "location: [1564:1567](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['O-linked (Glc...) serine. /evidence=ECO:0000269|PubMed:30127001; propagated from UniProtKB/Swiss-Prot (P46531.4); glycosylation site']\n",
      "\n",
      "type: misc_feature\n",
      "location: [1603:1618](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['NOTCH1']\n",
      "    Key: gene_synonym, Value: ['AOS5; AOVD1; hN1; TAN1']\n",
      "    Key: note, Value: ['propagated from UniProtKB/Swiss-Prot (P46531.4); Region: Interaction with DLL4. /evidence=ECO:0000250|UniProtKB:Q07008']\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Bio.SeqFeature.SeqFeature"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Lets take a look at some of these features...\n",
    "\n",
    "for feature in record.features[:20]:\n",
    "    print(feature)\n",
    "    \n",
    "type(record.features[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example in the real\n",
    "Lets try to do a common exercise in sequence analysis, and compare two sequences for similarity.  To do this, we will introduce you to a dot plot, which is a visual way of comparing sequences to each other. Let's compare the sequence of the same gene between mouse and human."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pick a gene\n",
    "gene = \"Hoxc6\"\n",
    "\n",
    "\n",
    "# Search for human records\n",
    "handle = Entrez.esearch(\n",
    "                db = 'nucleotide',\n",
    "                term = f\"Homo[Orgn] AND {gene}[Gene]\"\n",
    "    )\n",
    "\n",
    "hum = Entrez.read(handle)\n",
    "\n",
    "humID = hum['IdList'][0]\n",
    "\n",
    "# Search for mouse records\n",
    "handle = Entrez.esearch(\n",
    "                db = 'nucleotide',\n",
    "                term = f\"Mus[Orgn] AND {gene}[Gene]\"\n",
    "    )\n",
    "\n",
    "mus = Entrez.read(handle)\n",
    "\n",
    "musID = mus['IdList'][0]\n",
    "\n",
    "#Retrieve full human record\n",
    "with Entrez.efetch(\n",
    "    db=\"nucleotide\", rettype=\"gb\", retmode=\"text\", id=humID\n",
    ") as handle:\n",
    "    hRecord = SeqIO.read(handle, \"gb\")\n",
    "    \n",
    "#Retrieve full mouse record\n",
    "with Entrez.efetch(\n",
    "    db=\"nucleotide\", rettype=\"gb\", retmode=\"text\", id=musID\n",
    ") as handle:\n",
    "    mRecord = SeqIO.read(handle, \"gb\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate the similarity betwen the two sequences over a certain window size\n",
    "window = 10\n",
    "hSeq = hRecord.seq\n",
    "mSeq = mRecord.seq\n",
    "data = [[(hSeq[i:i+window] == mSeq[j:j+window]) for j in range(len(hSeq)-window)] for i in range(len(mSeq)-window)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAqIAAAH/CAYAAACM+b9HAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/d3fzzAAAACXBIWXMAAAsTAAALEwEAmpwYAABkwElEQVR4nO3deZxcVZn/8c83CYEQlrBDCBiUxWGNEAPjsAmCyMiqKIuKyxgZxXUcgZ+i4DKjuACKykQ2RUURQUFkVVZZA9kICVvYkiAQEYGAQNLP749zCm6Kqu7qdFXd6urvm9d95da5T917aqH79LnnnEcRgZmZmZlZuw0ruwJmZmZmNjS5IWpmZmZmpXBD1MzMzMxK4YaomZmZmZXCDVEzMzMzK4UbomZmZmZWCjdEzayjSHpI0tvKroeZmbWeG6JmbVSrkSXpg5JubPF1t5d0vaTnJD0u6dN14sZLihxXif2DpL36ca1zJH29j5iQtDhfY4Gk70ka3s/XtLuk+f15To1zfDDX5XtV5Qfm8nMGcv5OU+89k3StpP8oo05mNrS5IWrW5SStDVwO/B+wFrApcGUfTxsTEasA2wFXARdJ+mCTq7ZdvsaewOHAR5t8/kY9ALxX0ohC2QeAe0uqj5nZkOGGqFmHkfQvuYfqaUmzJe2fy98iaZGkjfLj7XLMG/PjnSXdlMseLTQcPwdcERG/iIgXI+LZiJjTSF0i4q8RcSpwAvAtScP6qONk4AjgC7m385IGrjEXuAHYusZ7saKkUyQtzNspuWw0cBkwttB7O7aR11TDX4FZwNvzNdcE3gJcXFWX/fNrfTq/9n8pHAtJmxYev9IrLGnt3Kv8tKSnJN1QeB/HSvqtpCclPSjpU7UqKGknSX8t9hpLOkjSzLw/SdJUSc/kXuzv1TpPI2r10BdfX35tP5J0WX7f/yJp/fzZ/F3SXElvKjz3WEkPSHpW0t2SDqq+lqTv5Oc+KOkdy1t3Mxt83BA16yCSVgAuIfVYrgt8EviFpC0i4iZSr+ZPJY0CzgW+FBFzJW1Mapj9AFgHmABMz6fdCXgqN1KfkHRJju+PC3N9tuijjlOAXwAnRcQqEbFfA695S2AXYFqNw1/M9Z9A6p2dlF/zYuAdwMJ8nVUiYmE/X1PRz0i9oACHAr8HXizUcXPgPOAzpPf3j8AlkkY2cO7/Aubn560H/D8gcmP0EmAGsCGpZ/gzkt5efYKIuAVYDOxRKD4c+GXePxU4NSJWA94AnN9AvQbiPcCXgLVJ79PNwJ358QVAsSH8AOnzXR04Efi5pA0Kx3cE7snPPQk4U5JaXH8z6xBuiJq13+9y79jTkp4GflQ4thOwCvDNiHgpIv4M/AE4LB8/gfQL/TZgIfDDXH4EcHVEnBcRL0fE3yJiej42DjgS+DSwMfAgqVHVH5VG3poN1LFRd0r6O6kxdgZwdo2YI4CvRsQTEfEkqSHz/n5epxEXAbtLWp3UIP1Z1fH3ApdGxFUR8TLwHWAUqee0Ly8DGwCvy5/NDRERwJuBdSLiq/l9nAf8hNQQruU88nssaVVgX179HF8GNpW0dkQ8lxuu9Ywtfv/yd3DnBl5H0UURcUdE/JP03v0zIn4WEUuBXwOv9IhGxG8iYmFE9ETEr4H7SH9QVDwcET/Jz/0p6b1ar5/1MbNByg1Rs/Y7MCLGVDbg44VjY4FHI6KnUPYwqceM3Ag6h3Qb+7u5QQOwEannqZYXSA2H23PD4UTgLbnR1agN879P9VXHftg+ItaIiDdExJeqzlcxNp+7eJ2GbsFL2rhw2/653mIj4gXgUnIvX0T8pbd65Lo+SmOv+dvA/cCVkuZJOjaXv46qRiGpt7ReI+yXwMGSVgQOBu6MiEqdPgJsDsyVdLukd/ZSn4XF71/+DvZ3stzjhf0XajxepfJA0gckTS+8xq1JvZ8Vf63sRMTzeXcVzGxIGNF3iJm10UJgI0nDCg2zjckTZyRtCHyF1Hv4XUlvjogXSY2iSbVOCMwEovC4st+f258HAU/w6i3UunWsutZALSQ12GYXrlPpne31OhHxCP1r0PwM+DOpoV6rHttUHuRbxxsBC3LR88DKhfj1SbfjiYhnSbfn/0vSVsA1km4nfWYPRsRmjVQuIu6W9DBpSELxtjwRcR9wWL7dfzBwgaS18hCG/lpcfC2S1l+Oc1Se+zpSL++ewM0RsVTSdPr33TOzLuYeUbPOciupIfAFSStI2h3YD/hVbvycA5xJ6gF7DPhaft4vgLdJeo+kEZLWkjQhHzsbOEjShDy+83jgxoh4uq/KSFpP0tGkxu9xueFZt475aY8Dr1/ud2BZ5wFfkrSO0uz/LwM/L1xnrX727PbmOmAv0jjbaucD/y5pz/we/hdpbORN+fh04HBJwyXtA+xWeaKkd0raNH9+zwBL83Yb8IykYySNys/dWtKbe6njL4FPAbsCvylc432S1smfz9O5eGk/X3/FDGCr/H1ZiTQcZHmNJv3B8GSu54eoMSnNzIYuN0TNOkhEvATsT+r1WkQaP/qBPLP8U6TbtsfnW/IfAj4kaZfc+7cvqYH0FKlhtF0+559Jt3wvJfVqbkrqUevN05IWk2aT7wscEhFnNVBHSA3lLfOt2N8N6A2BrwNTSb26s0gTYr6e6zGX1FCdl6+1vLPmyeeLiPhTRDxV49g9wPtIjdRFpIb3fvm9gDT+dj9SI/AI4HeFp28GXA08R5rU86OIuDaPidyPNBHrwXzeM0hjgOs5D9gd+HNELCqU7wPMzkMQTgUOzcMw+i0i7gW+mut8H/2/bV88193Ad0mv+3FSr3L1sAczG8L06hAzMzMzM7P2cY+omZmZmZVi0DREJe0j6R5J9xdmnZqZmZlZG0g6K69HfVed45L0/dxWmylp+77OOSgaokrZRH5IGpO2JWl26Jbl1srMzMxsSDmHNCa9nneQxsVvBkwGftzXCQdFQ5S0LM39ETEvTw74FXBAyXUyMzMzGzIi4nrShNh6DgB+lid/3gKMqcqk9hqDZR3RDUlr7lXMJ6WFq2vEyA09C8vMzMyaZslLCzpiDdyXF81rSRtn5Dpv+BipJ7NiSk7d3Kha7bUNScsN1jRYGqK1PvjXfAiSJpPfQA1fnWHDRre6XmZmZmZdITc6+9PwrNZQe61osNyan0/KYlIxjlezq7wiIqZExMSImOhGqJmZmVlbNdReKxosDdHbgc0kbSJpJHAocHHJdTIzMzNrv56lrdkG7mLgA3n2/E7APyKi7m15GCS35iNiSU4zeAUwHDgrImb38TQzMzOz7hM9pVxWUiW729qS5pPSP68AEBGnA38kZeO7H3ielAGw93N2a2YlT1YyMzOzZuqYyUqP39OSNs4K623R9tc3KHpEzczMzCzrKadHtBUGyxhRMzMzM+syHdUjKmkl4HpgRVLdLoiIr0j6NbBFDhsDPB0RE0qppJmZmVmJoqQxoq3QUQ1R4EVgj4h4TtIKwI2SLouI91YCJH0X+EdpNTQzMzMrUxfdmu+ohmikmVPP5Ycr5O2VAbmSBLwH2KP9tTMzMzOzZuq4MaKShkuaDjwBXBURtxYO7wI8HhH31XnuZElTJU3t6VnchtqamZmZtVn0tGYrQcc1RCNiaR7/OQ6YJGnrwuHDgPN6ea4zK5mZmZkNEh11a74oIp6WdC2wD3CXpBHAwcAOpVbMzMzMrEzNyYLUETqqR1TSOpLG5P1RwNuAufnw24C5ETG/pOqZmZmZWRN1Wo/oBsBPJQ0nNZLPj4g/5GOH0stteTMzM7Mhwcs3tUZEzATeVOfYB5txjRcW3sCosbs041RmZmZm7ddFyzd11K35dhg1dhdeWHhD2dUwMzMzG/I6qiEqaQtJ0wvbM5I+Uzj+eUkhae2BXMc9omZmZjZYRfS0ZCtDp92avweYAGk9UWABcFF+vBGwF/BIWfUzMzMzs+bpqIZolT2BByLi4fz4ZOALwO/Lq5KZmZlZyTxGtC1emSUvaX9gQUTM6O0JA8ms5HGjZmZmNih0UWYlpfTunUXSSGAhsBXwLHANsHdE/EPSQ8DEiFjU2zlGjNyw816YmZmZDVpLXlqgsusA8OK9N7akjbPi5ju3/fV16q35dwB3RsTjkrYBNgFmSIKU+vNOSZMi4q9lVtLMzMys7boos1KnNkRfySkfEbOAdSsHGu0RNTMzM7PO1nFjRCWtTJodf2HZdfG4UTMzM+s4XTRGtON6RCPieWCtXo6Pb1ddvN6omZmZdRzPmh9a3DNqZmZm1nylNEQlnSXpCUl3FcoOkTRbUo+kiYXytSRdI+k5SaeVUV+nBTUzM7OO0UW35svqET0H2Keq7C7gYOD6qvJ/AscDn299terzbXozMzOz5ipljGhEXC9pfFXZHIC8RFOxfDFwo6RN21ZBMzMzs07VRWNEO26y0kBImgxMBtDw1Rk2bHTJNTIzMzNrrojuWUe0qyYrRcSUiJgYERPb0Qj1uFEzMzOz5ddVPaLt5nGjZmZm1nYlTSxqha7qETUzMzOzwaOUHlFJ5wG7A2tLmg98BXgK+AGwDnCppOkR8fYc/xCwGjBS0oHA3hFxdwlVNzMzMyuXJysNTEQcVufQRXXix7euNs31wsIbfMvezMzMrAG+Nd9kXvzezMzMWsoL2g9MrcxKhWOflxSS1s6PV5D0U0mzJM2RdFz7a9w/7hE1MzOzlulZ2pqtBJ2UWQlJGwF7AY8Uig8BVoyIbYAdgI9VL4bfqdwzamZmZlZfKQ3RiLieNDmp2snAF4AohgOjJY0ARgEvAc+0vJJN4J5RMzMzazrfmm8+SfsDCyJiRtWhC4DFwGOkntLvREStRiySJkuaKmlqT8/i1lbYzMzMzAakIxa0l7Qy8EVg7xqHJwFLgbHAGsANkq6OiHnVgRExBZgCMGLkhlF93MzMzGzQ66LlmzqlR/QNwCbAjLxm6DjgTknrA4cDl0fEyxHxBPAXYGJpNR0gjxs1MzOzAfGt+eaKiFkRsW5EjM9rhs4Hto+Iv5Jux++hZDSwEzC3xOoOiMeNmpmZmSVlLd90HnAzsIWk+ZI+0kv4D4FVgLuA24GzI2JmG6ppZmZm1nl6elqzlaDTMitVjo8v7D9HWsLJzMzMzLpIR9yaH+o8btTMzMwa1kU9oh2TWUnSBEm3SJqel2CalMvHS3ohl0+XdHoZdW4lpwU1MzOzRkUsbclWhk7KrHQScGJETAC+nB9XPBARE/J2VHuq2F6exGRmZmZDTVljRK+vkaYzgNXy/urAwrZWyszMzGww6KJ1RDtiQfvsM8AVkr5D6ql9S+HYJpKmkVJ7fikiat7HljQZmAyg4aszbNjo1tbYzMzMzJZbJ01W+k/gsxGxEfBZ4Mxc/hiwcUS8Cfgc8EtJq9U6QURMiYiJETGxGxqhHjdqZmZmr+EF7VviSODCvP8bUmpPIuLFiPhb3r8DeADYvJQatpnHjZqZmVk366Rb8wuB3YBrgT2A+wAkrQM8FRFLJb0e2Ax4TZ55MzMzsyHBY0QHJmdW2h1YW9J84CvAR4FTJY0A/kke6wnsCnxV0hJgKXBURDzV/lqbmZmZdYCSbqO3QqdlVtqhRuxvgd+2tkZmZmZm1m6ddGvezMzMzPrSRbfm2z5ZSdJGkq6RNEfSbEmfzuXfljRX0kxJF0kak8v3knSHpFn53z3aXWczMzMza74yZs0vAf4rIv4F2An4hKQtgauArSNiW+Be4LgcvwjYLyK2Ic2sP7eEOpuZmZl1hi5avqntt+Yj4jHS2qBExLOS5gAbRsSVhbBbgHfnmGmF8tnASpJWjIgX21VnMzMzs47hW/PNkdN8vgm4terQh4HLajzlXcA0N0LNzMzMBr/SJitJWoU0G/4zEfFMofyLpNv3v6iK3wr4FrB3L+d0ik8zMzPrbu4RHRhJK5Aaob+IiAsL5UcC7wSOiIgolI8DLgI+EBEP1Dtvt6X4XB5OC2pmZmaDRRmz5kXKIz8nIr5XKN8HOAbYPyKeL5SPAS4FjouIv7S5uoPOqLG7uDFqZmbWzTxZaUD+DXg/MEvS9Fz2/4DvAysCV6W2KrdExFHA0cCmwPGSjs/xe0fEE22t9SDiHPVmZmZdrItuzZcxa/5GQDUO/bFO/NeBr7e0UmZmZmbWds6sZGZmZjaYdFGu+VKXbzIzMzOzoauTUnyeIGmBpOl52zeXj5R0dk7xOUPS7u2us5mZmVnH6OlpzVaCMm7NV1J83ilpVeAOSVflYydHxHeq4j8KEBHbSFoXuEzSmyO6qF+6jV5YeIMnM5mZmVlH6JgUn708ZUvgTzn+CUlPAxOB21pc1a7kRqiZmdkg10V9cZ2W4vNoSTMlnSVpjVw2AzhA0ghJmwA7ABvVOd9kSVMlTe3pWdzq6puZmZm1Xxfdmi+tIVojxeePgTcAE0g9pt/NoWcB84GpwCnATaTb+6/hzEpmZmZmg0cpyzfVSvEZEY8Xjv8E+EMuXwJ8tnDsJuC+tla4y3ncqJmZ2SDSRQvad1KKzw0KYQcBd+XylSWNzvt7AUsi4u42VrnrOS2omZmZlaGTUnweJmkCEMBDwMfysXWBKyT1AAvycztKN/QoDvb6m5mZDRkRZdegaQZDis+HgC1aWaeBciPOzMzM2sa35s3MzMzMBqaTMittJ+nmnEHpEkmrFZ6zbT42Ox9fqd31NjMzM+sIXr5pQCqZlf4F2An4hKQtgTOAYyNiG+Ai4L8BJI0Afg4cFRFbAbsDL5dQbzMzMzNrorY3RCPisYi4M+8/C1QyK20BXJ/DrgLelff3BmZGxIz8nL9FxNL21npo84x6MzOzDhI9rdlK0EmZle4C9s+HDuHV7EmbAyHpCkl3SvpCL+dzZqUW8GQsMzOzDuJb8wNXI7PSh0m36e8AVgVeyqEjgJ2BI/K/B0nas9Y5nVnJzMzMbPAopSFaJ7PS3IjYOyJ2AM4DHsjh84HrImJRRDxPWuZp+zLqbWZmZla6iNZsfZC0j6R7JN0v6dgax1fPE85n5AnmH+rrnJ2UWWnd/O8w4EvA6fnQFcC2OcPSCGA3wJmVSuZxo2ZmZkOHpOHAD4F3AFuSEhFtWRX2CeDuiNiONLn8u5JG9nbeMnpEK5mV9pA0PW/7kl7QvcBcYCFwNkBE/B34HnA7MB24MyIuLaHeVuC0oGZmZiUpZ4zoJOD+iJgXES8BvwIOqIoJYNXc6bgK8BRptaS6OimzEsCpdZ7zc9ISTtZBPInJzMyse0iaDEwuFE2JiCl5f0Pg0cKx+cCOVac4DbiY1KG4KvDeiN6n45eRa97MzMzMlleLZrjnRueUOodrdSJWDyx9O+nu9R7AG4CrJN2QJ6XXVMYY0ZUk3VYYyHpiLv+apJn5Vv2Vksbm8kmFW/gzJB3U7jqbmZmZdYxy1hGdz6tLawKMI/V8Fn0IuDCS+4EHgTf2dtIyxoi+COyRB7JOAPaRtBPw7YjYNiImAH8Avpzj7wIm5vJ9gP/Lk5asA3ncqJmZWVe6HdhM0iZ5AtKhpNvwRY8AewJIWo+UrGhebyctY4xoAM/lhyvkLaq6bUeTu3vzkk0VK/HabmDrIB43amZm1lrR0/6mUEQskXQ0aTWj4cBZETFb0lH5+OnA14BzJM0i3co/JiIW9XbeUnoW8xIAdwCbAj+MiFtz+TeADwD/AN5aiN8ROAt4HfD+iKg5A6s4yFbDV8eL2pfnhYU3uFFqZmbWRSLij6T13Itlpxf2F5JSszeslAXtI2JpvtU+Dpgkaetc/sWI2Aj4BXB0If7WiNgKeDNwnKSV6pzXmZU6hBuhZmZmLeIUn80REU8D15LGfhb9EnhXjfg5wGJg61bXzZrH40bNzMyaqJzJSi1Rxqz5dSSNyfujgLcBcyVtVgjbn7SwPXlQ7Ii8/zrSwNeH2llnGxj3jpqZmVktZYwR3QD4aR4nOgw4PyL+IOm3krYAeoCHgaNy/M7AsZJezsc+3tfAV+tMHjdqZmbWBCVMVmqVMmbNzwTeVKP8Nbfic/m5wLmtrpe1XiUtqBujZmZmBs6sZG3mRqiZmdkAlTSxqBXa3hDNM96vB1bM178gIr4i6dek8Z8AY4Cn88x6JB0HfARYCnwqIq5od73NzMzMOoIbogNSyaz0nKQVgBslXRYR760ESPouaS1RJG1JWr1/K2AscLWkzSNiaQl1NzMzM7Mmafus+Zx/9DWZlSrHJQl4D3BeLjoA+FVEvBgRDwL3A5PaWGVrMS/vZGZm1g8RrdlKUMo6opKGS5oOPAFcVcmslO0CPB4R9+XHGwKPFo7Pz2W1zjtZ0lRJU3t6Freg5tYKHjdqZmY2NJUyWSnfVp+Q1xO9SNLWEXFXPnwYr/aGQspV+ppT1DnvFGAKwIiRG3bP2gZmZmZmFV00RrSjMivlhesPBn5dCJsPbFR4PA5Y2J4ampmZmVmrdExmpXz4bcDciJhfeMrFwKGSVpS0CbAZcFsbq2wl8LhRMzOzOnqiNVsJOiazUj52KMveliciZks6H7gbWAJ8wjPmu5/HjZqZmdVRUl74VuiYzEr52AfrlH8D+EYLq2UdypmYzMzMulepY0TN+lJJC2pmZmZZF92aL60hmpdwmibpD/nxmpKuknRf/neNXH6EpOmFrUfShLLqbe3nHlEzM7PuVGaP6KeBOYXHxwJ/iojNgD/lx0TELyJiQk73+X7goYiY3ua6mpmZmXWE6OlpyVaGsha0Hwf8O3BGofgA4Kd5/6fAgTWeWr3GqJmZmdnQ4lvzA3YK8AWg2PxeLyIeA8j/rlvjee+ll4aoMysNHR43amZmNviVsY7oO4EnIuKOfj5vR+D5Qgam14iIKRExMSImDhs2eqBVtQ7mcaNmZjZkRU9rthKUsY7ovwH7S9oXWAlYTdLPgcclbRARj0nagJSHvug1a4yamZmZ2eDV9h7RiDguIsZFxHhS4/LPEfE+UgalI3PYkcDvK8+RNAw4BPhVm6trZmZm1lk8RrQlvgnsJek+YK/8uGJXYH5EzCulZjYoeNyomZkNCT09rdlKUMat+VdExLXAtXn/b8CevcTt1K562eBUWfze40fNzMwGh1IbombN5kaomZl1vZJuo7dCx2RWKpR/XlJIWruqfGNJz0n6fHtraoORb9ObmZl1vjJ7RCuZlVarFEjaiDQ+9JEa8ScDl7WnajbYuWfUzMy6VklLLbVCJ2VWgtTY/AIQVfEHAvOA2e2on5mZmZm1XsdkVpK0P7AgImYUAyWNBo4BTmxnBc3MzMw6kpdvWn61MitJWhn4IvDlGk85ETg5Ip5r4NxO8Wl1edyomZl1g+jpaclWho7IrAScC2wCzJAEMA64U9IkYEfg3ZJOAsYAPZL+GRGnVZ84IqYAUwBGjNywe6aUWVN43KiZmVlnaXtDNCKOA44DkLQ78PmIeFcxRtJDwMSIWATsUig/AXiuViPUzMzMbEjw8k1mZmZmZgPTMZmVqsrH14k/oaUVsiHHmZjMzGzQcY+oWXeopAU1MzMbNKKnNVsJ3BC1Ic89omZmZuXomBSfkiZIukXS9LwE06Rcvpaka3J6T09SMjMzs6HN64g2RSXFZ8VJwIkRMYG0nuhJufyfwPGAc8ybmZmZdZFOSvEZvJp3fnVgIUBELI6IG0kNUjMzM7MhLXqiJVsZypo1fwopxeeqhbLPAFdI+g6pgfyW/p5U0mRgMoCGr86wYaMHXFEzMzOzjuJZ88uvVorP7D+Bz0bERsBngTP7e+6ImBIREyNiohuhZmZmZp2tI1J8Svo5sB9p3CjAb1j2tr2ZmZmZAZSUF74V2t4jGhHHRcS4vGj9ocCfI+J9pDGhu+WwPYD72l03MzMzM2ufUjMrVfkocKqkEaSJSZMrB3Lu+dWAkZIOBPaOiLvLqKSZmZlZqbpojGjHpPjMM+N3qBM3vm2VMuuD04KamZk1hzMrmfWT04KamVmpvKD9wEh6SNKsShalXHaIpNmSeiRNLMSuIOmnOX6OpOPKqLNZkXtEzcysLBHRkq0MZd6af2tELCo8vgs4GPi/qrhDgBUjYhtJKwN3SzovIh5qUz3NzMzMrAU6ZrJSRMwBkPSaQ8DoPIlpFPAS8Ex7a2dmZmbWIbposlJZY0QDuFLSHTkbUm8uABYDjwGPAN+JiKdqBUqaLGmqpKk9PYubW2MzMzMza6qyekT/LSIWSloXuErS3Ii4vk7sJGApMBZYA7hB0tURMa86MCKmAFMARozcsHv+XDAzMzOrcI/owETEwvzvE8BFpMZmPYcDl0fEyzn+L8DEXuLNzMzMulb0REu2MpSRa360pFUr+8DepIlK9TwC7KFkNLATMLf1NTUzMzOzViqjR3Q94EZJM4DbgEsj4nJJB0maD/wrcKmkK3L8D4FVSI3V24GzI2JmCfU2MzMzK18XrSPa9jGieWzndjXKLyLdpq8uf460hJOZmZmZdRFnVjJrA2diMjOzpulp0VaCTsqsdIKkBblsuqR9c/l4SS8Uyk8vo85mA+G0oGZm1izdNFmpkzIrAZwcEd+pEftARExoQ53MWsZpQc3MzJblW/NmJXDvqJmZLbcumqzUaZmVjpY0U9JZktYolG8iaZqk6yTV7VZyZiUbLNw7amZmBopofwtY0thiZiXgk8A9wCJSI/VrwAYR8WFJKwKrRMTfJO0A/A7YKiJ6zTfvzEpmZmbWTEteWqCy6wDw9Hvf2pI2zphfX9P219cxmZUi4vGIWBoRPcBPyNmWIuLFiPhb3r8DeADYvIx6m5mZmVnzdExmJUkbFMIOImdbkrSOpOF5//XAZsBr8sybmZmZDQWeNT8w6wEXSapc/5c5s9K5kiaQbs0/BHwsx+8KfFXSEmApcFREPNX2WpuZmZl1gpLW/GyFTsqs9P468b8FftvqepmZmZlZe5W5jqiZmZmZ9VNZt9FboZMyK20n6eZcfomk1XL5XnmZp1n53z3KqLOZmZmZNVeZC9q/NSImRMTE/PgM4NiI2IY0k/6/c/kiYL9cfiRwbvuralYOL3xvZmav4VzzLbEFcH3evwp4F0BETKss9wTMBlbKa4uadT0vfG9mZtWipzVbGTops9JdwP55/xBgoxrPexcwLSJerHVSZ1aybubeUTMz6zZlTVb6t2JmJUlzgQ8D35f0ZeBi4KXiEyRtBXyLtO5oTRExBZgCzqxk3ce9o2ZmBnTV8k2dlFlpbkTsHRE7AOeRMigBIGlcjvtARDxQ65xmZmZmNrh0UmaldXPZMOBLwOn58RjgUuC4iPhLu+trZmZm1kk8RnRg1gNulDQDuA24NCIuBw6TdC8wF1gInJ3jjwY2BY7Pyz1NrzRazczMzIacLpo1r4juHErpMaJmZmbWTEteWqCy6wCw6O27taSNs/YV17X99TmzkpmZmdkgUtZt9FYoK7PSGEkXSJoraY6kf5X07fx4pqSL8thQJI2UdHbOrDRD0u5l1Nms03l5JzMzG2zKWkf0VODyiHgjsB0wh7SI/dYRsS1wL3Bcjv0oQM6stBfw3TyhycwKvLyTmdnQUNZkJUn7SLpH0v2Sjq0Ts3uezzNb0nV9nbOMWfOrAbsCZwJExEsR8XREXBkRS3LYLcC4vL8l8Kcc+wTwNDARM6vJPaNmZt2tjIaopOHAD4F3kNpmh0nasipmDPAjYP+I2IqUoKhXZfQsvh54Ejhb0jRJZ+RlnIo+DFyW92cAB0gaIWkTYAdqZ11yZiUz3DNqZmYtMQm4PyLmRcRLwK+AA6piDgcujIhH4JUOxF6V0RAdAWwP/Dgi3gQsBl7p3pX0RWAJ8ItcdBYwH5gKnALclI+/RkRMiYiJETFx2LDqtq3Z0OPeUTOzLhRqyVbs0Mvb5MJVNwQeLTyen8uKNgfWkHRtTuP+gb5eShmz5ucD8yPi1vz4AnJDVNKRwDuBPSOvK5Vv13+28mRJNwH3tbXGZoOUe0fNzKxRxVTpNdRa2ql6GakRpDvXewKjgJsl3RIR99a7Ztt7RCPir8CjkrbIRXsCd0vaBziGNK7g+Uq8pJUrt+4l7QUsiYi7211vs8HMPaNmZt2jpMlK81l2aOQ4UgKi6pjLI2JxRCwCridNSq+rrHVEPwn8QtJIYB7wIeB2YEXgKkkAt0TEUcC6wBWSeoAFwPvLqbLZ4DVq7C68sPAG95Camdnyuh3YLM/XWQAcShoTWvR74DRJI4CRwI7Ayb2dtJSGaERM57Uz3zetE/sQsEWtY2bWODdCzcy6Q/S0P8FTRCyRdDRwBTAcOCsiZks6Kh8/PSLmSLocmElKGnpGRNzV23md4tPMzMysAZ2S4nPhW97akjbO2Juuafvr65jMSrn8k3mh1NmSTirEbyvp5lw+S9JKZdTbzMzMzJqnrDGilcxK787jRFeW9FbSelTbRsSLktYFyOMMfg68PyJmSFoLeLmkepuZmZmVKqIjOmabou0N0UJmpQ9CyqwEvCTpP4FvRsSLubyyCOrewMyImJHL/9buOpuZmZlZ83VSZqXNgV0k3SrpOklvzvGbAyHpCkl3SvpCCXU2MzMz6whl5ZpvhTJuzVcyK30yIm6VdCppQfsRwBrATsCbgfMlvT6X75zLngf+JOmOiPhT9YlzBoDJABq+Os6uZGZmZt2mjFnzrVJGj2itzErb5/ILI7mNNO1/7Vx+XUQsygvd/zHHv4ZTfJoNnBe/NzOzdumYzErA74A9ACRtTloIdRFpvaptc4alEcBuOd7MWsDrjZqZdbaI1mxlqHtrXtKOwJyIeEbSKNLt8+1JjcD/iYh/DOC6tTIrLQbOknQX8BJwZM43/3dJ3yOt6B/AHyPi0gFc28zMzMw6QN0F7SXNBrbLK+lPIY3PvIDUg7ldRBzcvmr2nxe0NzMzs2bqlAXtH97+bS1p47zuzqvb/vp6m6w0LCKW5P2JEVEZl3mjpOmtrZaZmZmZ1TJUJivdJelDeX+GpInwyvhNLyhvZmZmZgPSW0P0P4DdJD0AbAncLGke8JN8bLlI2kLS9ML2jKTPSFpT0lWS7sv/rpHjJxViZ0g6aHmvbWZmZjbYddNkpbpjRF8JkFYlLUI/grTs0uNNu7g0HFgA7Ah8AngqIr4p6VhgjYg4RtLKwEt5rOoGwAxgbGHYQE0eI2pmZmbN1CljRB/cbq+WtHE2mXFVR40RrdgKmB0Rz8IrDdMtC+uADsSewAMR8bCkA4Ddc/lPgWuBY/LaoRUrkWbOm5mZmQ1JQ2WMaMWPgecKjxfnsmY4FDgv768XEY8B5H/XrQRJ2jHP4p8FHFWvN1TSZElTJU3t6VncpCqamZmZWSs00hBVFO7fR0QPTUgNmtcQ3R/4TV+xEXFrRGxFSvN5nKSV6sQ5s5KZmZl1tQi1ZCtDIw3ReZI+JWmFvH2atAj9QL0DuLMw5vTxPAaU/O8T1U+IiDmkHtmtm3B9MzMzs0EnelqzlaGRhuhRwFtIk4rmkyYWTW7CtQ/j1dvyABcDR+b9I4HfA0jaJKf2RNLrgC2Ah5pwfTMzMzMrUSO32J+JiEOLBZLWHshF80z4vYCPFYq/CZwv6SPAI8AhuXxn4FhJLwM9wMcjYtFArm9mZmY2WPWUdBu9FRrpEb1d0k6VB5LeBdw0kItGxPMRsVYxX31E/C0i9oyIzfK/T+XycyNiq4iYEBHbR8TvBnJtM2utFxbeUHYVzMxskGikR/Rw4CxJ1wJjgbWAPVpZKTMbvEaN3aXsKpiZdbWyJha1Qp89ohExC/gGaazoW4GjI2L+8l6wXmalwvHPS4ri7X9Jx0m6X9I9kt6+vNc2MzMzG+yiRy3ZytBnj6ikM4E3ANsCmwOXSDotIn64PBeMiHuACfnclcxKF+XHG5HGjj5SuP6WpPVGtyL1yF4tafOIWLo81zczMzOzztDIGNG7gLdGxIMRcQWwE7B9k67/Smal/Phk4Assmz3pAOBXEfFiRDwI3A9MatL1zawNPG7UzKx5uinXfJ89ohFxsqTtJFUGft0QER9p0vVfyawkaX9gQUTMkJbpHt4QuKXweH4uew1Jk8lLS2n46nhRe7PO4HGjZmZWS589opI+BfyClHJzXeDnkj450AsXMyvl5Zy+CHy5VmiNsprtdmdWMjMzs243pMaIAv8B7BgRiwEkfQu4GfjBAK/9SmYlSdsAmwCV3tBxwJ2SJpF6QDcqPG8csHCA1zYzMzMblIbaOqICihODllK7l7K/XsmsFBGzImLdiBgfEeNJjc/tI+KvpIxLh0paUdImwGbAbU24vpmZmZmVqJEe0bOBWyVdlB8fCJw5kIvWyaxUU0TMlnQ+cDewBPiEZ8ybdYcXFt7g8aNmZv3UTeuIKhqYJiVpe1KqTQHXR8S0VldsoEaM3LCk+V9mZmbWjZa8tKAjWoCzNtmvJW2cbR68pO2vr26PqKQ1Cw8fytsrxyopOM3MzMysfcpaaqkVers1fwdpdrqADXh1gpBy+euX54KStgB+XSh6PWm2/L8CW+SyMcDTETFB0hHAfxfityWNH52+PNc3MzMzs85QtyEaEZtU9iVNi4g3NeOC9TIrRcQphet9F/hHjv8Fafko8uz637sRata9PG7UzKx33TRrvpHJSlBn3c4mqM6shNL6Te8B9qgR/8pMezPrTm6Empn1rpsmKzWyfFMrvZJZqWAX4PGIuK9G/HtrxL9C0mRJUyVN7elZ3MRqmlm7OS2omVn3622y0ucKD9etekxEfG8gFy5kVjqu6lDNXk9JOwLPR8Rd9c4ZEVOAKeBZ82aDnXtGzcxqGyqTlVYt7P+k6nEzvJJZqVIgaQRwMLBDjfhavadmZmZmNkj1NlnpxBZfu1bP59uAuRExv1goaRhwCLBri+tkZmZm1tGG4mSlpuols1K9Xs9dgfkRMa/VdTMzMzPrZN00WamUhmhEPA+sVaP8g3XirwV2am2tzMzMzKydSmmImpmZmdny6aZb830u3yRpPUlnSrosP95S0kcGclFJn5U0W9Jdks6TtJKkQ3JZj6SJNZ6zsaTnJH1+INc2MzMzs87QyDqi5wBXAGPz43uBzyzvBSVtCHwKmBgRWwPDSWND7yLNmL++zlNPBi5b3uuamZmZdYNo0VaGRm7Nrx0R50s6DiAilkha2oTrjpL0MrAysDAi5gCkxErLknQgMA/wKvVmZmY2pA2pW/PAYklrkRvLknYi54FfHhGxAPgO8AjwGPCPiLiyXryk0cAxQJ/LSTmzkpmZmdng0UhD9HPAxcAbJP0F+BnwyeW9oKQ1gAOATUi3+0dLel8vTzkRODkinuvr3BExJSImRsTEYcNGL28VzczMzDpWhFqylaHPW/MRcaek3YAtAAH3RMTLA7jm24AHI+JJAEkXAm8Bfl4nfkfg3ZJOAsYAPZL+GRGnDaAOZmZmZlayRmbNHwKMiojZwIHAryVtP4BrPgLsJGllpQGhewJz6gVHxC4RMT4ixgOnAP/jRqiZmZkNVT0t2srQyK354yPiWUk7A28Hfgr8eHkvGBG3AhcAdwKzch2mSDpI0nzgX4FLJV2xvNcwMzMzs86niN4n7EuaFhFvkvS/wKyI+GWlrD1VXD4jRm5Y1koEZmZm1oWWvLSgI6arX7/+IS1p4+z619+0/fU1snzTAkn/Rxrb+S1JK9JYT6qZmZmZNVlPF3W1NdKgfA9pQft9IuJpYE3gvwdy0TqZlU6QtEDS9Lztm2PXknRNzqrksaFmZmZmXaKhBe2BqZDSbOayuct7wUJmpS0j4gVJ55MyK0Fapuk7VU/5J3A8sHXezMzMzIasHjpihEBTNNIQvZS0mL2AlUjrf94DbDXA6y6TWQkYXyswIhYDN0radADXMzMzM7MO0+et+YjYJiK2zf9uBkwCblzeC/aRWeloSTMlnZUXvu8XZ1Yys968sPCGsqtgZjZggVqylaHfk44i4k7gzct7wV4yK/0YeAMwgdRA/e5y1M2ZlcysrlFjd3Fj1MwGvW5aR7TPW/OSPld4OAzYHnhyANesmVkpIl7JrCTpJ8AfBnANM7OaRo3dpewqmJlZ1sgY0VUL+0tIY0Z/O4BrvpJZCXiBlFlpqqQNIuKxHHMQcNcArmFmZmbWlcq6jd4KjeSaPxFA0qrpYTw3kAtGxK2SKpmVlgDTgCnAGZImkCZGPQR8rPIcSQ8BqwEjJR0I7B0Rdw+kHmZmZmZWrkZuzW8NnEtaPxRJi4AjI2K5eywj4ivAV6qK399L/PjlvZaZmZlZNylrPGcrNHJrfgrwuYi4BkDS7rnsLa2rlpmZmZnV0k0N0UZmzY+uNEIBIuJaYEBT0utkVpog6ZacVWmqpEk5dgVJP5U0S9IcSccN5NpmZmZm1hka6RGdJ+l40u15gPcBDy7vBXvJrHQ4cGJEXJbTe54E7A4cAqwYEdvkCU53SzovIh5a3jqYmZmZDVbdNFmpkR7RDwPrABcCF+X9Dw3wupXMSiN4NbNSkCYkAayey8jlo3PsKOAl4JkBXt/MzMzMStbIrPm/k3owmyIiFkiqZFZ6AbgyIq6U9ChwRT42jFfHoF5AWgD/MVKj9bMR8VSz6mNmZmY2mPR0T4do/YaopIt7e2JE7L88F6zKrPQ08JucWWkSqZH5W0nvAc4kLX4/CVhKysK0BnCDpKsjYl6Nc08GJgNo+Oo4u5KZmZlZ5+qtR/RfgUeB84BboWkDEmpmVgKOAD6dY34DnJH3Dwcuj4iXgSck/QWYCLymIRoRU0gz+hkxcsNoUn3NzMzMOkbPEBkjuj7w/4CtgVOBvYBFEXFdRFw3gGu+kllJkkiZleaQxoTulmP2AO4rxO+hZDSwEzB3ANc3M+s356g3s04RLdrKULdHNCKWApcDl0taETgMuFbSVyPiB8t7wV4yK00DTs2Tkv5JvsUO/BA4m5TyU8DZETFzea9vZrY8nKPezKz5ep2slBug/05qhI4Hvk+aPT8gdTIr3QjsUCP2OdISTmZmZmZDXjctaN/bZKWfkm7LX0Za33O5U3qamZmZmVXrrUf0/cBiYHPgU2k4J5Buj0dErFbviWZmZmbWGj0aApOVImJYRKyat9UK26oDbYRK+nRO7zlb0mdy2bclzZU0U9JFksbk8vGSXsipP6dLOn0g1zYzMzMbzLppslIjmZWaStLWwEdJ64NuB7xT0mbAVcDWEbEtcC9QzCn/QERMyNtR7a6zmZmZmTVf2xuiwL8At0TE8xGxBLgOOCgirsyPAW4BxpVQNzMzM7OO1tOirQxlNETvAnaVtJaklYF9gY2qYj5MmiRVsYmkaZKuk1R3DRVJkyVNlTS1p2dx82tuZmZmZk3TZ675ZouIOZK+RboV/xwwg7SeKACSvpgf/yIXPQZsHBF/k7QD8DtJW0XEMzXO7cxKZmZm1tW6Kdd8GT2iRMSZEbF9ROwKPEXOoiTpSOCdwBERETn2xYj4W96/A3iANJPfzMzMbMjpQS3ZylBKQ1TSuvnfjYGDgfMk7QMcA+wfEc8XYteRNDzvvx7YjBp55s3MzMysdSTtI+keSfdLOraXuDdLWirp3X2ds+235rPfSloLeBn4RET8XdJpwIrAVXnN0lvyDPldga9KWgIsBY6KiKdKqreZmZlZqcoYe5g7BX8I7AXMB26XdHFE3F0j7lvAFY2ct5SGaES8ZsJRRGxaJ/a3wG9bXikzsxZ5YeENzlVvZoPdJOD+iJgHIOlXwAHA3VVxnyS1297cyElLuTVvZjaUuBFqZs3Uo9ZsxdWH8ja5cNkNgUcLj+fnsldI2hA4CGg4+VBZY0Rfk1kpl38yjz2YLemkXLaXpDskzcr/7lFGnc3MzMy6WURMiYiJhW1K4XCt2UzVowROAY6JiKWNXrPtt+arMiu9BFwu6VLSAvYHANtGxIuVCU3AImC/iFiYn3sFVS1wMzMzs6GipMXn57Psuu/jgIVVMROBX+W5PmsD+0paEhG/q3fSMsaIvpJZCUDSdaRu3InANyPiRYCIeCL/O63w3NnASpJWrMSZmZmZDSUlLZR+O7CZpE2ABcChwOHFgIjYpLIv6RzgD701QqGzMittDuwi6dacQanWINd3AdPqNUKdWcnMzMys+XIa9qNJd6bnAOdHxGxJR0k6annP20mZlUYAawA7kWZanS/p9ZWF7SVtRVoOYO9ezu3MSmZmZtbVysqsFBF/BP5YVVZzYlJEfLCRc3ZSZqX5wIWR3EYaArE2gKRxwEXAByLigTLqbGZmZmbNVco6opLWjYgnCpmV/pXU8NwDuFbS5sBIYJGkMcClwHER8Zcy6mtmZmbWKUqarNQSnZRZ6SzgLEl3kWbTHxkRIeloYFPgeEnH5+fvXZnMZGZmZjaUuCE6QHUyK70EvK9G+deBr7ejXmZmZmbWPmX1iJqZmZnZcoiSJiu1QsdkVpK0naSbcwalSyStlstHSjo7l8+QtHsZdTYzMzOz5mp7Q7Qqs9J2wDslbQacARwbEduQZsj/d37KRwFy+V7AdyWV0oA2MzMzK1tPi7YylNGgeyWzUl4ctZJZaQvg+hxzFWnxeoAtgT/BK9mWniZlYTIzMzMbctwQHZh6mZXuAvbPMYfwaj7TGcABkkbktFI7sGyu01c4s5KZmZnZ4NH2hmhEzCFlSLoKuJxXMyt9GPiEpDuAVUlLOAGcRVrsfipwCnBTjq917ikRMTEiJg4bNrqVL8PMzMysFNGirQxlLd90JnAmgKT/AeZHxFxy+s68oP2/59glwGcrz5V0EykTk5mZmZkNYh2TWalQNgz4EnB6jl0ZUEQslrQXsCQi7i6j3mZmZmZlKyvXfCt0UmalT0v6RD5+IXB23l8XuEJSD7AAeH/7q2tmZmZmzdZJmZVOBU6tUf4QaUa9mZmZ2ZDnFJ9mZmZmVopuaoi2bNa8pLMkPSHprkLZmpKuknRf/neNXD5J0vS8zZB0UOE5IyVNkXSvpLmS3lXremZmZmY2uLRy+aZzgH2qyo4F/hQRm5EWqT82l98FTIyICfk5/yep0lv7ReCJiNictLj9dS2ss5mZmVlH8/JNDYiI6yWNryo+ANg97/8UuBY4JiKeL8SsxLLvx4eBN+Zz9gCLWlBdMzMzM2uzdi9ov15EPAaQ/123ckDSjpJmA7OAoyJiiaQx+fDXJN0p6TeS1qt3cmdWMjMzs27Xo9ZsZSgjxWdNEXFrRGwFvBk4TtJKpB7bccBfImJ74GbgO72cw5mVzMzMrKs51/zye1zSBgD53yeqA3IK0MXA1sDfgOeBi/Lh3wDbt6eqZmZmZtZK7W6IXgwcmfePBH4PIGmTyuQkSa8jrRv6UEQEcAmvjivdE3BWJTMzMxuyPFmpAZLOIzUg15Y0H/gK8E3gfEkfAR4BDsnhOwPHSnqZ1Dv88YioTEo6BjhX0inAk8CHWlVnMzMzM2ufVs6aP6zOoT1rxJ4LnFvnPA8DuzaxamZmZmaDVk9p/ZfN58xKZmZmZoOIMys1oJ+ZlY4oZFaaLqlH0oR87PKcbWm2pNMlDW9Vnc3MzMysfTois1JE/CIiJuTMSu8nTVSanp/znojYjjSLfh1eHVdqZmZmNuR002SlljVEI+J64Kmq4gNIGZXI/x5Y46mHAecVzvNM3h0BjKS898rMzMzMmqhjMisVvJdCQxRA0hWkNUefBS5odSXNzMzMOpUXtG8RSTsCz0fEXcXyiHg7sAGwIrBHL893ik8zMzOzQaLTMisdSlVvaEVE/JO0IP4B9U7uFJ9mZmbW7ZxrfvnVzKwEIGkYaSLSrwplqxQariOAfYG5bautmZmZWYfpIVqylaFTMitBWrR+fkTMK5SNBi6WtCIwHPgzcHqr6mxmZmZm7dMRmZVy/LXATlVljwNvbm7NzMzMzAavblo+qKMmK5mZmZnZ0NHuzEqH5AxJPZImFsrXknSNpOcknVZ1nh0kzZJ0v6TvSyppOK2ZmZlZ+bx8U2PO4bWZle4CDgauryr/J3A88Pka5/kxMBnYLG/V5zQzMzMbMrppslJbMytFxJyIuKdG7OKIuJHUIH1FnjG/WkTcHBEB/Iza2ZjMzMzMbJBp2WSlJtkQmF94PD+XmZmZmQ1JnqzUPrXGg9Z9/51ZyczMzGzw6PQe0fnAuMLjccDCesERMQWYAjBi5Ibd9AeDmZmZGVDexKJW6Oge0Yh4DHhW0k55tvwHKGRjMjMzMxtqummyUrszKz0F/ABYB7hU0vSIeHuOfwhYDRgp6UBg74i4G/hP0gz8UcBleTMzMzOzQa6MzEoX1YkfX6d8KrB1k6plZmZmNqh109jDjr41b2ZmZmbdq9MnK5mZmZlZgScrNaA/KT4LxzfOaT4/nx+vKml6YVsk6ZRW1dnMzMys00WL/itDp6T4rDiZwmSkiHg2IiZUNuBh4MLmV9XMzMzM2q2Vk5WulzS+qmwOQFqJaVl5pvw8oOZK9JI2A9YFbmhyVc3MzMwGDd+abzJJo4FjgBN7CTsM+HXOOV/vPM6sZGZmZjZIdMpkpROBkyPiuVq9pdmhwPt7O4kzK5mZmVm3K2vx+VbolIbojsC7JZ0EjAF6JP0zIk4DkLQdMCIi7iixjmZmZmbWRB3REI2IXSr7kk4Anqs0QrPDgPPaXS8zMzOzTtM9/aEdlOKzD+8B9m1VXc3MzMwGC9+ab0B/U3wWnndCjbLXN6NOZmZmZtY5OuLWvJmZmZk1xss3NaA/mZUkrSDpp5JmSZoj6bga57u4eC4zMzMzG9w6JbPSIcCKEbENsAPwseJi+JIOBp5rWU3NzMzMBgmn+GxARFxPmpxULJsTEffUCgdGSxoBjAJeAp4BkLQK8Dng662qq5mZmdlg0dOirQwdkVkJuICU2vMx4BHgOxFRacR+Dfgu8HxfJ3FmJTMzM7PBo1MmK00ClgJjgTWAGyRdDawGbBoRn63OW1+LMyuZmZlZtyvrNnordEpD9HDg8oh4GXhC0l+AicBawA6SHiLVdV1J10bE7qXV1MzMzMyaolNuzT8C7KFkNLATMDcifhwRYyNiPLAzcK8boWZmZjaUddMY0U7JrPRD4GzSrHoBZ0fEzFbVzczMzGyw6gnfmu9TfzIrRcRzpCWcejvfQ8DWA6+ZmZmZmXWCThkjamZmZmYN6J7+0PZnVvq2pLmSZkq6SNKYXL6XpDtyZqU7JO1ReM43JD0qyQvam5mZmXWRdmdWugrYOiK2Be4FKqk8FwH75cxKRwLnFp5zCWl5JzMzM7Mhr4doyVaGdmdWujIiluSHtwDjcvm0iFiYy2cDK0laMR+7JSIea1U9zczMzKwcZY4R/TDw6xrl7wKmRcSL/T2hpMnAZAANX51hw0YPrIZmZmZmHcYL2g+QpC8CS4BfVJVvBXwL2Ht5zuvMSmZmZtbtylrzsxXa3hCVdCTwTmDPiFcXwpI0jrS00wci4oF218vMzMzM2qutDVFJ+wDHALtFxPOF8jHApcBxEfGXdtbJzMzMbDApa2JRK7Ry+abzgJuBLSTNl/QR4DRgVeAqSdMlnZ7DjwY2BY7P5dMlrZvPc1LOzLRyPs8JraqzmZmZmbWPoovSRBV5jKiZ2dD1wsIbGDV2l7KrYV1myUsLVHYdAN79uv1b0sa54OGL2/76nFnJzMy6jhuh1s26abJSp2RWGi/phcJt+dMLzzksZ1yaKelySWu3qs5mZmZm1j6dklkJ4IGImJC3owAkjQBOBd6anzOTNJ7UzMzMbEiKiJZsZeiIzEq9UN5GSxKwGrCw96eYmZmZWbNJ2kfSPZLul3RsjeNH5DvYMyXdJGm7vs7Zyh7RvnwYuKzweBNJ0yRdJ2kXgIh4GfhPYBapAbolcGa9E0qaLGmqpKk9PYtbWHUzMzOzcpSRa17ScOCHwDtI7bHDJG1ZFfYgaYnObYGvkZMM9aaUhmiNzEqPARtHxJuAzwG/lLSapBVIDdE3AWNJt+aPq3FKIGVWioiJETHR6T3NzMysG/W0aOvDJOD+iJgXES8BvwIOKAZExE0R8ff8sJE73+1viBYyKx1RyawUES9GxN/y/h3AA8DmwIRc9kCOPR94S7vrbGZmZtbtineW8za5cHhD4NHC4/m5rJ6PsOyd75o6JbPSOsBTEbFU0uuBzYB5wErAlpLWiYgngb2AOe2ss5mZmVkniRZlVoqIKdS/nV5rjdGaFZH0VlJDdOe+rtmyhmjOrLQ7sHbOjPQV0m31FUmZlQBuyTPkdwW+KmkJsBQ4KiKeyuc5Ebhe0svAw8AHW1VnMzMzM6tpPrBR4fE4akwgl7QtcAbwjsrd7t44s5KZmZlZAzols9K+G+/bkjbOHx/5Y93Xl5fUvBfYE1gA3A4cHhGzCzEbA38GPhARNzVyTWdWMjMzM7NeRcQSSUcDVwDDgbMiYrako/Lx04EvA2sBP8p3vpdExMTeztuyHlFJZ5EmJT0REVvnsm8D+wEvkSYkfSginpY0Evg/YCJp4tanI+La/JxrgQ2AF/Kp946IJ/q6vntEzczMrJk6pUf0HRu9oyVtnMsevaztr69TMit9FCAitiFNSPqupGLdjihkXeqzEWpmZmbWrUpavqklOiWz0pbAn3LME8DTpN5RMzMzM+tSnZJZaQZwgKQRkjYBdmDZmVlnS5ou6fic6tPMzMxsSIoW/VeGTsmsdBZpWYCpwCnATfk4pNvy2wC75O39vZzXKT7NzKypXlh4Q9lVMOtabZ81X8istGchs9IS4LOFmJuA+/KxBfnfZyX9kpRi6me1zl1ciNWTlczMrBlGjd2l7CqYLaOvvPCDSadkVlqZNIN/saS9SNP9785rVo2JiEU57/w7gavbWWczMzOzTtJNa8B3SmaldYErJPWQFkmt3H5fMZevQFqz6mrgJ62qs5mZmZm1T8saohFxWI3iM+vEPgRsUaN8MWnikpmZmZnRXbfmy5w1b2ZmZmZDWMsaopLOkvSEpLsKZV+TNDMvxXSlpLG5fFIumy5phqSDcvnKki6VNFfSbEnfbFV9zczMzAYDL9/UmHN4bWalb0fEthExAfgDKScpwF3AxFy+D/B/eaISwHci4o3Am4B/k/SOFtbZzMzMrKP1RLRkK0O7Mys9U3g4Gqgs3/R8IePSSlXl1+T9l4A7eTUbk5mZmZkNYmWsI/oN4APAP4C3Fsp3JC1s/zrg/YWGaeX4GGA/4NS2VdbMzMysw3TPVKUSJitFxBcjYiNSVqWjC+W3RsRWwJuB4yStVDmWb9OfB3w/IubVO7czK5mZmZkNHmXOmv8l8K7qwoiYAywGti4UTwHui4hTejthREyJiIkRMXHYsNHNrKuZmZlZR+ghWrKVoa0NUUmbFR7uD8zN5ZtUJidJeh1pTdGH8uOvA6sDn2lnXc3MzMystdqdWWlfSVsAPcDDwFE5fGfgWEkv52Mfz2k9xwFfJDVY78zZmE6LiDNaVW8zMzOzTtZNC9qrm/KVFo0YuWF3vjAzMzMrxZKXFqjsOgDsNHb3lrRxbll4bdtfnzMrmZmZtcELC28ouwpmHccNUTMzszYYNXYXN0atKTxZqQH9SfGZj20r6eacynNWZfkmSe/Nz5kt6aRW1dfMzKzVRo3dpewqmHWUjkjxmWfM/xw4Kq8lujvwsqS1gG8De+by9STt2cI6m5mZmXU055pvQH9SfAJ7AzMjYkaO+1tELAVeD9wbEU/muKupsfaomZmZ2VARES3ZytApKT43B0LSFcA6wK8i4iTgfuCNksYD84EDgZG9nHsyMBlAw1fHi9qbmZmZda5OSfE5grSW6BH534Mk7RkRfwf+E/g1cANpkfslrznpq+d2ZiUzMzPrap6s1BzFFJ/zgesiYlFEPA/8EdgeICIuiYgdI+JfgXuA+0qprZmZWQt5Rr0NRR2R4hO4AthW0sp54tJuwN35Oevmf9cAPg44q5KZmXUdz6i3RnmMaAP6k+IzIv4u6XvA7aQJTH+MiEvzqU6VtF3e/2pE3NuqOpuZmZl1Oqf4HASc4tPMzMyaqVNSfG63/lta0saZ8deb2v762j5r3szMzMyWX1lrfrZCWzMrFY59XlJIWrtQdpyk+yXdI+nthfKRkqZIulfSXEleR9TMzMysC7SyR/Qc4DTgZ8VCSRsBewGPFMq2BA4FtgLGAldL2jwvav9F4ImI2FzSMGDNFtbZzMzMrKP1dNGwyrZmVspOBr4Ay/QrH0BaxP7FiHiQtJD9pHzsw8D/5nP2RMSiVtXZzMzMzNqn3cs37Q8sqKTyLNgQeLTweD6woaQx+fHXJN0p6TeS1uvl/JMlTZU0tadncVPrbmZmZtYJnGt+OUhamXSb/cu1DtcoC9LQgXHAXyJie+Bm4Dv1ruHMSmZmNhR48fuhrSeiJVsZ2tkj+gZgE2CGpIdIDcw7Ja1P6gHdqBA7DlgI/A14Hrgol/+GnHHJzMxsqPLi99Yt2tYQjYhZEbFuRIyPiPGkxuf2EfFX4GLgUEkrStoE2Ay4LdIip5eQFsYH2JOcccnMzGyoc8/o0NRNt+bbmlkpIs6sFRsRsyWdT2pkLgE+kWfMAxwDnCvpFOBJ4EOtqrOZmdlg4p5RG+ycWcnMzKwLvLDwBjdMW6xTMittvs7ElrRx7n1yqjMrmZmZWf+5ETp0OLNSA/qbWSmXbyzpOUmfL5RdLmmGpNmSTpc0vFV1NjMzG+w8btQGk1ZOVjoH2Ke6sFZmpYKTgcuqyt4TEdsBWwPrAIc0t5pmZmbdY9TYXdwY7XJevqkB/cyshKQDgXnA7KrzPJN3RwAjq59nZmZmy/JtehssOiKzkqTRpNnxJ9Z53hXAE8CzwAW9nN+ZlczMzKyrddPyTZ2SWelE4OSIeK7WcyPi7cAGwIrAHvWu4cxKZmZm1u0ielqylaGds+aLmZXg1cxKk4AdgXdLOgkYA/RI+mdEnFZ5ckT8U9LFwAHAVW2st5mZmZm1QNsaohExC1i38jin+ZwYEYuAXQrlJwDPRcRpklYBVo2IxySNAPYFPALbzMzMhqyeLpou08rlm84Dbga2kDRf0keW4zSjgYslzQRmkMaJnt7EapqZmZlZSZxZyczMzKwBnZJZaeM1t2lJG+eRp2a1/fW1dda8mZmZmVlFR2RWkrSCpJ9KmiVpjqTjCrE75PL7JX1feaaTmZmZ2VDUQ7RkK0OnZFY6BFgxIrYBdgA+Jml8PvZjYDKwWd5ec04zMzOzoSIiWrKVoVMyKwUwOs+MHwW8BDwjaQNgtYi4OdI79DPgwFbV2czMzMzapyMyK5GyJS0GHiP1lH4nIp4CNgTmF+Lm57J653dmJTMzswFyrvrO1k255tu2jmghs9LeNQ5PApYCY4E1gBskXQ3UGg9a952KiCnAFPCseTMzs+XlXPXWLp2SWelw4PKIeBl4QtJfgImkxevHFc4xDljYxjqbmZmZdZSy8sK3QttuzUfErIhYNyLGR8R40m327SPir6Tb8XsoGQ3sBMyNiMeAZyXtlGfLfwD4fbvqbGZmZtZpPFmpAf3MrPRDYBXgLuB24OyImJmP/SdwBnA/8ABwWavqbGZmZrV53Ki1gjMrmZmZmTWgUzIrrbP6Fi1p4zz5j3ucWcnMzMzMhoa2ZlaSdIKkBZKm523fXL6WpGskPSfptEL8qoXY6ZIWSTqlVXU2MzOzvvk2fbm6aYxoK2fNnwOcRlqEvujkiPhOVdk/geOBrfMGQEQ8C0yoPJZ0B3BhC+pqZmZmDRo1dhdeWHiDl3kqSVlrfrZCGZmVasUujogbSQ3SmiRtBqxLWtLJzMzMSuRGqDVDGWNEj5Y0M9+6X6MfzzsM+HV06+wqMzMzswZ00635djdEf0xa2H4CKZ3nd/vx3EOB83oLcIpPMzMzs8GjnZmViIjHK/uSfgL8oZHnSdoOGBERd/Rxfqf4NDMzs67W48xKy0fSBoWHB5EWsG/EYfTRG2pmZmZmg0vLekRzZqXdgbUlzQe+AuwuaQIQwEPAxwrxDwGrASMlHQjsHRF358PvAfZtVV3NzMyseTyjvrW6abqMMyuZmZmZNaBTMiutsvImLWnjPPf8g86sZGZmZmZDQ1snK5mZmZnZwIQnK/Wtnyk+x0t6oVB+eo3zXVw8l5mZmZkNbp2S4hPggYiYUOtEkg4Gnmtq7czMzMwGIaf4bEB/Unz2RtIqwOeArw+4UmZmZmaDnDMrDUy9FJ+bSJom6TpJxTUfvkbKwPR8Xyd2ZiUzMzOzwaNTUnw+BmwcEW8i9X7+UtJqec3RTSPiokZOHhFTImJiREwcNmx00ytvZmZmVrZo0X9l6IgUnxHxIvBi3r9D0gPA5sCbgR3yYvcjgHUlXRsRu7ez3mZmZmbWfG1tiEraICIeyw9fSfEpaR3gqYhYKun1wGbAvIiYSupFRdJ44A9uhJqZmdlQ1k3JiDolxeeuwFclLQGWAkdFxIAnOpmZmZl1m7IaopL2AU4FhgNnRMQ3q44rH9+XNLfngxFxZ6/n7KZWdZFTfJqZmVkzdUqKzxVa1MZ5uZfXJ2k4cC+wFzAfuB04LCLuLsTsC3yS1BDdETg1Inbs7ZpO8WlmZmY2iESLtj5MAu6PiHkR8RLwK+CAqpgDgJ9FcgswRtIGvb+YFq1FVfYGTB5KcYOhjp0eNxjq2C1xg6GOnR43GOrY6XGDoY7dEjcY6tif19KtGzAZmFrYJheOvZt0O77y+P3AaVXP/wOwc+Hxn4CJvV6z7Bfdwjdz6lCKGwx17PS4wVDHbokbDHXs9LjBUMdOjxsMdeyWuMFQx/68lqG4AYfUaIj+oCrm0hoN0R16O69vzZuZmZlZX+YDGxUejwMWLkfMMtwQNTMzM7O+3A5sJmkTSSOBQ4GLq2IuBj6gZCfgH/Hqsp01tXUd0TabMsTiyrx2t8SVee2hFlfmtbslrsxrd0tcmdceanFlXrsVr2XIiYglko4GriAt33RWRMyWdFQ+fjrwR9KM+ftJyzd9qK/zdu3yTWZmZmbW2Xxr3szMzMxK4YaomZmZmZXCDVEzMzMzK0VXTFaS9EbSav4bkpIDLAQujog5pVbMzMzMzOoa9D2iko4hpZkScBtpeQEB50k6tsy6lU3Stv2M31jSmLw/XtK7JW1dJ3aipIMk7Zf/EKgVI0k7Sjo4x+4oqeE8vbXOK2mFGmVrVz0eJmlY3h8paXtJazZwvY83ELNKPt+YqvKRxdcm6a2S/kvSO2qco+HPZah/Jjm+5Z/LUPxMzMw6Qtkr9Tdhpf97gRVqlI8E7uvHeS4r7K8G/C9wLnB4VdyPCvvrAz8GfgisBZwAzALOBzbo43ofb7BeqwDbA2NqvD4VHr8V+C/gHYWypaQlFL4GbNnHdY4FHgTmAv+R/z0TmA18rhC3Gynt19XA30npvP4CXAtsVIjbO1/7MuCMvF2ey/Zu8LU/UvX65gNPAlcC4wvH7izsHwg8DjxG6iW/Ffhzfu5+hbjPVW3/BSyqPK7zee8MPAJcAzwK7Fs4NgNYI+//N3AT8CXgKuB/q15XQ5/LUPtMyvxchtpn4s2bN2+dspVegQG/gPSL4HU1yl8H3FNVtn2dbQfgsULcb4Fv5l+gF+fHK+ZjxV+wlwOfzL+cZgLHABvnst8X4hr65Zpjm/YLFpgGbA18I/9im5HrOr7G+zUbGEVqUD8LrJPLRwN3FeKmFY5tAlyU9/cCrizEzalznU2AOYXH36+z/QB4phB3O7BV3n83cB+wU6VOVfVbP1/nGWCLwvdhaiHuWeDXwJeBr+Tt75X9Qlzx874G2D7vv77qfMX3aCowKu+PAGZWvQcNfS5D7TMp83MZap9JfizgPaS0fQL2zOf8ODCsuk5Vz/1zjbK1qx6/L59vMsv+0XwQsGbeXwf4GekP+F8D4wpx3wP+rbd6FGLXzN+Z/8iv5Yukxv+3yT8nC7FvBU4Dfs+rP+s3rXHOt5M6Gi7OsT8G9mmkPvn5X65xvo/U+E59eLB8Jv35XLrlM/HW+q30Cgz4BcA+vNqjMCVvlR6Ffapil5J6Yq6psb1QiJte9bwvknoz1mLZX4DTCvuPVD1nemG/oV+uObZpv2Cp6gEBJuUfIo8CN1UdqzxnOPBE8Yde1bWKv7yHV9V3dmH/PmBEjc9rJHB/1XszGTiyxraoEDej6jxbAfeQfoDW+0zuqnpOMW5j4ALgW8DKuWxejfoWn3NH1bHitW4Cts77l/PqHwkr9VaP3j6XofaZlPm5DLXPJB/7UX6vLwZ+DvwG+ABpqNOpxddStc0CXqw8rvOZfIm06PWR+bwnF47dXdj/NfBZUhrADwJXFY49SfrZ9jBwEvCm6vepEPvH/J35ManX+QfALsBXWbZT4JvA2aQG2QWkRtFHSX84HFKIOyWf81BSh8DOef+Pxfemt41le6r/B7g+n/cB4JN13reO/kz687l0y2firfVb6RVoyotIY113At5F6gXYCRheI+4uYLM653i0sD+Hqr8+8/+8s4GHC2UzCvtfr4ov/jBo6JdrLm/aL1iqekAKzxWwW1XZOcAvSX9lnkcalnAE6bbj+YW4s3LZ4fkH1vdy+crA3ELccfkHyTE59vC8Pw04rhD3Z+Atder5YGF/KrB+1fFxwHTg2eJ7VPnsgEmF8uFUNYJy+QGkPzLeXeszIWWGqPygf7bwPg9j2YbHtqRetJ/l7YH8Xk3ltcM7GvpchupnUsbnMtQ+k1w+K/+7AvA3YGR+PKJyLD+uNIreSOrFHk9qoL+Owt0olv35dCcwunD+4vnuKexX/4ybXn0+YDPgeNLP37mkP+A3r/W8/Hkt6OWcxXqMAP6S99eo+t7c28v34b7C42fqbM8CS4rXJf/BAYwhNZ5OrvG+dfRn0p/PpVs+E2+t30qvQFtfbPqltkWdYwcW9k8C3lYjZp+qL/xXgVVqxG0KXFCjvNdfrjmmab9gqWoA9fHejAAOI/2FOQJ4C+lWyRcqP7xy3Aqk20Snkf5qHZ7LR1E1RAL4F9LtzR/k+GOpGn9Hun2zcgP1exuwXY3y1YEvFh6/GVipRtx44H11zr0y6a/w62sce13VVvnFsDZwcFXscOAdwKdJwy/eS9XY3v58Li36TLZsw2cyZqCfyXJ8LisM5HMZwGfyb4PxM8ll0wr7l1cdm171+CBS79H++XGtPw7mAm8iDXWq7pmdXtj/P9LPzlHAd8k/e0m3Z68rxL2mV4r0c+9/KfQW5/KZpIbLxsA/yLdaSXexir19M3j1FvTGwC2FY7OrzjepxvUnsWzD6RFgvTqfxTKdGzW+k2eSeiaL1+3oz6Q/n0u3fCbeWr+VXoG2v+D0pX1z3t+SNE5z36qYHYHV8/6o/D/oJaQezdWrYt8AfB44Nf8PfFR1TFV83V+u+XjTGz4DeK/WLekzKuW6/ajfWmXXYSht/j605rqk4Uy1/pBeH7itRvlo0nCFi4H5NY5fU7VtUKknyw4rWoE0sfORvPWQ/uj+JbBxIW5aP17LYaQJcY+T7oxdTRovvwCYXIh7L+mW8pX52v+ey9cBflmI2540qe7uHHsl6U7ZrcAOhbivU6NxlI99q7D/B6ruQhWe3zNYPpP+fC51PpOrB9tn4q31W+kVaOuLTbcObiH1Gv4v6XbXl0l/VRZ7cGbzapf9FNIYkp3z8y8sxH0q/8/wJdKt8h+RJjvcDexe9uvt4724rOrxmjW2h0h/0a5ZiLszv97XN+PaNa65Vh/XfUMf516dNOZoLunW1t/yD6tvUmikU3vFg5lUrXiQn7d23p8IzCONP36YZW/ZrkL6g2U26a//J/N37YM16tjoqgz7FPbHkP5an0n65bBe4dhE0i+ZnwMbkX4B/4M0ceVNda57WIPXXX05rzuhTd/DVn4fXrMCRp3vw329fB/u6u37UOb3sJf3bDS9NP6B7YCj+nG+4dTpzc2fT80GNTUaZA1cp/Jze0R+n16zekn+Tk2kgT/a8+ezQ45fvz/1qTrPKPI4/hrHNhwsn0l/P5du/ky8NW8rvQJtfbHpF8twUq/kM8BquXwUy47pLM5WrR7gP736fHl/ZeDavL8xy95iaajhkR8PpPHxNIXGBw2uEpBje0jL0hS3l/O/8wpxDwLfIf3lehtpUPvYGu91Q9duwXWvII2xW79Qtn4uK06EaHTFg+Itn2t4tTd9c5btUfg9aWD/OFIv+/GkMVQ/Bf6nqo6NrspQ3D+D9Jf66/Jr/13h2G2knvHDSOPE3p3L9wRu7oDrlvk97OjvQ1nXrXqPai1/t3at2Hys5pJyhZ99Y/L+eNIwpK3rnavG89/Yx/Gay94B2/bjGiLd9TqYdHt7RwozyPtz7VZclzQMqzKmemR+r9esiulz+b6q+In5mvv19h7397tQ5vehzO+ht+ZupVegrS+2akmZqmPTC/u/AT6U988GJub9zYHbC3GzePUX+RoUBnmz7HjOhhoA1Y8ZYCOABlcJyLGfJ/1S3KZQ9mCN97BYv11IvcB/zecs3m5pdIWCZl/3nurn1jpG4ysezOXVv+hvqYorNg6qx1/dnv8dRmFySvX58+N6qzLc2ctzinXs7bVM64Drlvk97OjvQ8nfw0bXgW10SbmG1ljtbWPZWc2fo/Fl7xpdB7ahdVtrXPtzta7dguseSGPrIDe6Pu5uNLaebcPrz5b4fSjlut5av5Vegba+2PQ/dWXWenHZldVZ9gfv6qTZsQ/k57xMuhV2HYWJAKSxmTNJt+/n8mrjdR0KY0BpsAGQjzWtEUCDqwQUysaRGuHfA1alj2VzCmXDSRO5zi6UNXztJl/3StLEkWLv8XqknqarC2WNrnjwyXzOPUi3TU8BdgVOBM4txN0E7Jz39wOuKByrXs+20VUZ5vPqL+B5LNsDUqzjzaRfdIeQbtUemMt3Y9nesrKu24rv4bQaZYPu+9Di6+7fx/ew0XVgG11SrtE1Vr9fZ6teE7U/y95No7F1YBtdt7XR9Wybfd1pNLYOcn/Wx21kPdv+rD9b1vehlOt6a/1WegXa+mJzL2SN8rUp9MAUylcljb3Zgfqz8LbK/+P2dguhoQZALm9aI4AGVwmocWw/0riyv9Y49qsG3+t+X7tJ112DNKlsLukXx9/z+38Sy44x/CoNrnhA6i34NemH+izSMh8fo3AbK39PbiONy7uRV3+BrAN8qup8ja7K8JWqrfIDdX3gZ4W4CaRb0JeRlnM5Nb/u2RQWnm7Bdbdr8Lplfg+rvw9P5e/Dtwbwfdi9xvdhcp3vw9P5+7B5re/Dcl73zga+h71eN5c1ug5so0vKNbrGaqNroi7Xsnf5cb11YBtdt7Xf69k26brF97O3dZAbXR+30fVs+7P+bFnfh1Ku6631W+kVGAobDTYActlAGgFPkxoBbynEvZF0u36V6mvXqM8rsaS/KLeuFdvoOWlghYIacduQbjH1FVf3fDWed25fMTnuZyXF7ZxfS68pHfsRtwvpD5m9q8obWg2iKm7lBuMq5/tDjbhPUbgN2Efdl4ktfg97i+vlfCNJv1jelh8fQZoY9AmWbcCtSFo4vBJ3OGkppeq4kVVx9c5Xfd2a58vHGlp5I8f9N6nX5nu9xG1aiOvtfI2uA9voknLn0Ngaqw2tiVooO4C+l72bVqe8eh3YhtZtbfTazb4uDa65S+Pr4za6nm1/1p8t5ftQ9vfQW+s25TfdSiLpQxFxdjNjK3GSPkX6xTeH1Gv26Yj4fY65MyK2LzynoVhJnwSObiDuK6QxrCNI45Z2JI1LehvpduE36sRNIg2B6Cuu3vkurvGW7EH6oUNE7F8nTqSez77iGj1fzbgce1tETMr7HyW97xeRergviYhv1oj7D9L73lfcR0nrV/6uRtxs0tCSJZKmkH6wX0D6o2K7iDh4gHGLSeOfq+P+kY89QPrF8JuIeLLG+1Ur9vyIWNRAXM1zSvoF6TszitRbPTq/h3uS7jYcWRW3MukPulWAC3McEfHBOnF9na9y3eL5inGfAt5JWrljX9Iv/r+TeqI+HhHX5rhPA//eQFxD58uxbwOejIgZVe/ZGOAThf+nXlf1tj4WES9JWhvYNSIuzHEjSHdogvR92ZE0hv0R4IcRsTjHrQn8MyKep0GSViYNQ9gxInatcfzwiPhlg+fakjRsYUPS//fzgYsj4u7+XrvZ15X0ZtKY339WPXc8acjFzwtlw0n/j29O+q7NJ/0sfLoQswJpHdstSQ3XsyJiqaRRpFn4D+e4et+F1YGjK9+FXFb9fVgYES838H2YRGoQL9f3YQDfwwFd19qg7JbwUN/ox6DoRmMrcaS/HFfJ++NJf/V+Oj+eVvWchmL7GdfICgXNjruTtJLA7qRhCruTBv7vxrI9FNMajGvq+Wq8T7ez7BimWS2Ma3Q1iGbHTSP1WuxN6pV4knQ78Uhg1er3ppHYfsRVbtONIE0CqaxyoarvTVlxja680dQ4b9762ujH+r2NxvbnnCW9Zq8RXcI2DGs5STPrbLNIEyf6Hdtg3PCIeA4gIh4iNY7eIel7pF+IRY3GNhq3JCKWRvpr84GIeCY/5wXSEj2tipsI3EGaEPaPSD1AL0TEdRFxXSFuhwbjmn0+gGGS1pC0Fql37Mn8WhYDS1oYd5ekD+X9GZImAkjanDQhr1VxERE9EXFlRHwEGEua5b4PaRw0yxHbaNwwSSNJ471XJk1EhHQrfoUOiIPUWK0cWzW/uEdaHSdpdUnflDRX0t/yNieXjRlg3FP14noj6bJmxlXHSlpN0v9KOlfSYVVxP6oTd3gT4taX9GNJP5S0lqQT8s/r8yVt0EfcrAbj6p3vRw2cb82qbS3gtvxzZc2q19VQbI24NevE7VPYX13Smfm1/FLSenXixjQh7ptKvalImihpHnCrpIcl7Ya1T9kt4aGwkXpEJvDarEnjSbc1+h3bSBzptvCEqvOPII0pWlpV3lBsP+IaXaGgqXGF8srM69PopSe5jDjSAu3zyGtjksdmkW7fTm9h3Oo0thpEs+Om9fJ+jap63FBsP+I+m+v0MGlc6Z+An5B6Dr/SAXGfprGVN5oal8saXWN1oHHHVsVtX2erXmO4obh+nrPRtXSbHdfoerFlxTW0fm9/YvsR1+iShc2Oa2htXm+t30qvwFDYSLcOd65z7JfLE9tIHKlRVDPzBIVZzf2J7UdcQysUNDuuxvF/p85C3p0QV/WclYFNWh1HA6tBNDOOPHO7wfegodh+nnMsebF7UpKId1M7Z3VZcX2uvNGiuEbXWG12XKNrDPdn/dlGzzm96nn11tJtdty0wn5v68WWFdfQ+r39ie1HXKNLFjY7rqG1eb21fiu9At68efPmrf0bja+x2uy4htaVbTSun+dsdC3dZsf1tl5szYQE7YzLj/tcv7e/sY3E0fiShc2Oa2htXm+t3zxG1MxsaHovqefuujym8ynSShRrkmYdtyruBKj7u+eTyxHXn9hLSA2PV0TET0mNlpdaGPd7Savk41+qFEralLReZ9lxRMT8iDiE1It8FemOSk2NxjYY9xNSI3UVUirayrjN9UmrPrQkLiJ+APwPaS3eA0grWhwLLAA+hLVP2S1hb968efPWWRt5bOlgjxsMdezEOJZdR7rX8zUa259zdvJ74635m9cRNTOzZUh6JCI2Huxxg6GO3RI3GOrYitdsAzei7xAzM+s2kmbWO0TVUnGdHDcY6tgtcYOhjq14zdZaboiamQ1N6wFvJ2VeKhIpl/lgiRsMdeyWuMFQx1a8ZmshN0TNzIamP5CypE2vPiDp2kEUNxjq2C1xg6GOrXjN1kIeI2pmZmZmpfDyTWZmZmZWCjdEzczMzKwUboiaWWkkPVf1+IOSTiurPkWSHpK0duHx7pL+0ORr7C5puqTZkq5r5rnNzAYDT1YyMyuBpDHAj4B9IuIRSeuWXCUzs7Zzj6iZdSRJ50h6d+Hxc/nf3SVdJ+l8SfdK+qakIyTdJmmWpDfkuP0k3SppmqSrJa2Xy0+QdJakayXNk/Sp5ajbmpJ+J2mmpFskbStphKTbJe2eY/5X0jfy/j6S7pQ0Q9Kf8mkOBy6MiEcAIuKJ5X+3zMwGJ/eImlmZRkmaXni8JnBxA8/bDvgX4ClgHnBGREyS9GlSbvHPADcCO0VESPoP4AukHOAAbwTeSspJfY+kH0fEyzWuc42kpXl/FWBu3j8RmBYRB0raA/hZREyQ9EHggty43QfYUdI6pPzXu0bEg5LWzOfYHFghLxWzKnBqRPysgdduZtY13BA1szK9EBETKg9yQ25iA8+7PSIey895ALgyl88iNTABxgG/lrQBMBJ4sPD8SyPiReBFSU+QFreeX+M6b42IRfk6uwOfz+U7A+8CiIg/S1pL0uoRMVvSucAlwL9GxEuSdgKuj4gHc/xT+RwjgB2APUl5uG+WdEtE3NvA6zcz6wq+NW9mnWoJ+WeUJJEakxUvFvZ7Co97ePUP7B8Ap0XENsDHgJXqPH8p/f+jXDXKKosybwM8zatpAlU4VjQfuDwiFufG7vWknl4zsyHDDVEz61QPkXoMAQ4AVujn81cHFuT9I5tUp4rrgSPglZ7SRRHxjKSDgbWAXYHv5wlJNwO7Sdokx1duzf8e2CWPLV0Z2BGY0+R6mpl1NN+aN7NO9RPg95JuA/4ELO7n808AfiNpAXALsEkT63YCcLakmcDzwJF5qadvAntGxKN5GapTI+JISZOBCyUNA54A9oqIOZIuB2aSenLPiIi7mlhHM7OO5xSfZmZmZlYK35o3MzMzs1K4IWpmZmZmpXBD1MzMzMxK4YaomZmZmZXCDVEzMzMzK4UbomZmZmZWCjdEzczMzKwU/x/L9jRVRzK1fwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 864x576 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline \n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (12,8)\n",
    "\n",
    "ax = sns.heatmap(data)\n",
    "ax.set(xlabel=f'Human {gene}',ylabel=f'Mouse {gene}',title=f\"{gene} Dot Plot - Mouse vs Human\")\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
